// n8n-nodes-base/dist/nodes/Tapfiliate/Tapfiliate.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tapfiliate = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const AffiliateDescription_1 = require("./AffiliateDescription");
const AffiliateMetadataDescription_1 = require("./AffiliateMetadataDescription");
const GenericFunctions_1 = require("./GenericFunctions");
const ProgramAffiliateDescription_1 = require("./ProgramAffiliateDescription");
class Tapfiliate {
    description = {
        displayName: 'Tapfiliate',
        name: 'tapfiliate',
        icon: 'file:tapfiliate.svg',
        group: ['transform'],
        version: 1,
        subtitle: '={{$parameter["operation"] + ":" + $parameter["resource"]}}',
        description: 'Consume Tapfiliate API',
        defaults: {
            name: 'Tapfiliate',
        },
        usableAsTool: true,
        inputs: [n8n_workflow_1.NodeConnectionTypes.Main],
        outputs: [n8n_workflow_1.NodeConnectionTypes.Main],
        credentials: [
            {
                name: 'tapfiliateApi',
                required: true,
            },
        ],
        properties: [
            {
                displayName: 'Resource',
                name: 'resource',
                type: 'options',
                noDataExpression: true,
                options: [
                    {
                        name: 'Affiliate',
                        value: 'affiliate',
                    },
                    {
                        name: 'Affiliate Metadata',
                        value: 'affiliateMetadata',
                    },
                    {
                        name: 'Program Affiliate',
                        value: 'programAffiliate',
                    },
                ],
                default: 'affiliate',
                required: true,
            },
            ...AffiliateDescription_1.affiliateOperations,
            ...AffiliateDescription_1.affiliateFields,
            ...AffiliateMetadataDescription_1.affiliateMetadataOperations,
            ...AffiliateMetadataDescription_1.affiliateMetadataFields,
            ...ProgramAffiliateDescription_1.programAffiliateOperations,
            ...ProgramAffiliateDescription_1.programAffiliateFields,
        ],
    };
    methods = {
        loadOptions: {
            // Get custom fields to display to user so that they can select them easily
            async getPrograms() {
                const returnData = [];
                const programs = await GenericFunctions_1.tapfiliateApiRequestAllItems.call(this, 'GET', '/programs/');
                for (const program of programs) {
                    returnData.push({
                        name: program.title,
                        value: program.id,
                    });
                }
                return returnData;
            },
        },
    };
    async execute() {
        const items = this.getInputData();
        const length = items.length;
        const qs = {};
        let responseData;
        const returnData = [];
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        for (let i = 0; i < length; i++) {
            try {
                if (resource === 'affiliate') {
                    if (operation === 'create') {
                        //https://tapfiliate.com/docs/rest/#affiliates-affiliates-collection-post
                        const firstname = this.getNodeParameter('firstname', i);
                        const lastname = this.getNodeParameter('lastname', i);
                        const email = this.getNodeParameter('email', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            firstname,
                            lastname,
                            email,
                        };
                        Object.assign(body, additionalFields);
                        if (body.addressUi) {
                            body.address = body.addressUi.addressValues;
                            delete body.addressUi;
                            if (body.address.country) {
                                body.address.country = {
                                    code: body.address.country,
                                };
                            }
                        }
                        if (body.companyName) {
                            body.company = {
                                name: body.companyName,
                            };
                            delete body.companyName;
                        }
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'POST', '/affiliates/', body);
                        returnData.push(responseData);
                    }
                    if (operation === 'delete') {
                        //https://tapfiliate.com/docs/rest/#affiliates-affiliate-delete
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'DELETE', `/affiliates/${affiliateId}/`);
                        responseData = { success: true };
                    }
                    if (operation === 'get') {
                        //https://tapfiliate.com/docs/rest/#affiliates-affiliate-get
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'GET', `/affiliates/${affiliateId}/`);
                    }
                    if (operation === 'getAll') {
                        //https://tapfiliate.com/docs/rest/#affiliates-affiliates-collection-get
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const filters = this.getNodeParameter('filters', i);
                        Object.assign(qs, filters);
                        if (returnAll) {
                            responseData = await GenericFunctions_1.tapfiliateApiRequestAllItems.call(this, 'GET', '/affiliates/', {}, qs);
                        }
                        else {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'GET', '/affiliates/', {}, qs);
                            responseData = responseData.splice(0, limit);
                        }
                    }
                }
                if (resource === 'affiliateMetadata') {
                    if (operation === 'add') {
                        //https://tapfiliate.com/docs/rest/#affiliates-meta-data-key-put
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        const metadata = this.getNodeParameter('metadataUi', i)
                            ?.metadataValues || [];
                        if (metadata.length === 0) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Metadata cannot be empty.', {
                                itemIndex: i,
                            });
                        }
                        for (const { key, value } of metadata) {
                            await GenericFunctions_1.tapfiliateApiRequest.call(this, 'PUT', `/affiliates/${affiliateId}/meta-data/${key}/`, { value });
                        }
                        responseData = { success: true };
                    }
                    if (operation === 'remove') {
                        //https://tapfiliate.com/docs/rest/#affiliates-meta-data-key-delete
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        const key = this.getNodeParameter('key', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'DELETE', `/affiliates/${affiliateId}/meta-data/${key}/`);
                        responseData = { success: true };
                    }
                    if (operation === 'update') {
                        //https://tapfiliate.com/docs/rest/#affiliates-notes-collection-get
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        const key = this.getNodeParameter('key', i);
                        const value = this.getNodeParameter('value', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'PUT', `/affiliates/${affiliateId}/meta-data/`, { [key]: value });
                    }
                }
                if (resource === 'programAffiliate') {
                    if (operation === 'add') {
                        //https://tapfiliate.com/docs/rest/#programs-program-affiliates-collection-post
                        const programId = this.getNodeParameter('programId', i);
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            affiliate: {
                                id: affiliateId,
                            },
                        };
                        Object.assign(body, additionalFields);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'POST', `/programs/${programId}/affiliates/`, body);
                    }
                    if (operation === 'approve') {
                        //https://tapfiliate.com/docs/rest/#programs-approve-an-affiliate-for-a-program-put
                        const programId = this.getNodeParameter('programId', i);
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'PUT', `/programs/${programId}/affiliates/${affiliateId}/approved/`);
                    }
                    if (operation === 'disapprove') {
                        //https://tapfiliate.com/docs/rest/#programs-approve-an-affiliate-for-a-program-delete
                        const programId = this.getNodeParameter('programId', i);
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'DELETE', `/programs/${programId}/affiliates/${affiliateId}/approved/`);
                    }
                    if (operation === 'get') {
                        //https://tapfiliate.com/docs/rest/#programs-affiliate-in-program-get
                        const programId = this.getNodeParameter('programId', i);
                        const affiliateId = this.getNodeParameter('affiliateId', i);
                        responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'GET', `/programs/${programId}/affiliates/${affiliateId}/`);
                    }
                    if (operation === 'getAll') {
                        //https://tapfiliate.com/docs/rest/#programs-program-affiliates-collection-get
                        const programId = this.getNodeParameter('programId', i);
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const filters = this.getNodeParameter('filters', i);
                        Object.assign(qs, filters);
                        if (returnAll) {
                            responseData = await GenericFunctions_1.tapfiliateApiRequestAllItems.call(this, 'GET', `/programs/${programId}/affiliates/`, {}, qs);
                        }
                        else {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.tapfiliateApiRequest.call(this, 'GET', `/programs/${programId}/affiliates/`, {}, qs);
                            responseData = responseData.splice(0, limit);
                        }
                    }
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                    returnData.push(...executionErrorData);
                    continue;
                }
                throw error;
            }
        }
        return [returnData];
    }
}
exports.Tapfiliate = Tapfiliate;
