// @n8n/db/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthRolesService = exports.DbConnectionOptions = exports.DbConnection = exports.wrapMigration = exports.postgresMigrations = exports.mysqlMigrations = exports.sqliteMigrations = exports.CreateTable = exports.DslColumn = exports.NoUrl = exports.NoXss = exports.sqlite = exports.objectRetriever = exports.lowerCaser = exports.idStringifier = exports.sql = exports.separate = exports.isValidEmail = exports.isStringArray = exports.generateNanoId = exports.DateTimeColumn = exports.JsonColumn = exports.dbType = exports.datetimeColumnType = exports.jsonColumnType = exports.WithTimestampsAndStringId = exports.WithTimestamps = exports.WithStringId = void 0;
var abstract_entity_1 = require("./entities/abstract-entity");
Object.defineProperty(exports, "WithStringId", { enumerable: true, get: function () { return abstract_entity_1.WithStringId; } });
Object.defineProperty(exports, "WithTimestamps", { enumerable: true, get: function () { return abstract_entity_1.WithTimestamps; } });
Object.defineProperty(exports, "WithTimestampsAndStringId", { enumerable: true, get: function () { return abstract_entity_1.WithTimestampsAndStringId; } });
Object.defineProperty(exports, "jsonColumnType", { enumerable: true, get: function () { return abstract_entity_1.jsonColumnType; } });
Object.defineProperty(exports, "datetimeColumnType", { enumerable: true, get: function () { return abstract_entity_1.datetimeColumnType; } });
Object.defineProperty(exports, "dbType", { enumerable: true, get: function () { return abstract_entity_1.dbType; } });
Object.defineProperty(exports, "JsonColumn", { enumerable: true, get: function () { return abstract_entity_1.JsonColumn; } });
Object.defineProperty(exports, "DateTimeColumn", { enumerable: true, get: function () { return abstract_entity_1.DateTimeColumn; } });
var generators_1 = require("./utils/generators");
Object.defineProperty(exports, "generateNanoId", { enumerable: true, get: function () { return generators_1.generateNanoId; } });
var is_string_array_1 = require("./utils/is-string-array");
Object.defineProperty(exports, "isStringArray", { enumerable: true, get: function () { return is_string_array_1.isStringArray; } });
var is_valid_email_1 = require("./utils/is-valid-email");
Object.defineProperty(exports, "isValidEmail", { enumerable: true, get: function () { return is_valid_email_1.isValidEmail; } });
var separate_1 = require("./utils/separate");
Object.defineProperty(exports, "separate", { enumerable: true, get: function () { return separate_1.separate; } });
var sql_1 = require("./utils/sql");
Object.defineProperty(exports, "sql", { enumerable: true, get: function () { return sql_1.sql; } });
var transformers_1 = require("./utils/transformers");
Object.defineProperty(exports, "idStringifier", { enumerable: true, get: function () { return transformers_1.idStringifier; } });
Object.defineProperty(exports, "lowerCaser", { enumerable: true, get: function () { return transformers_1.lowerCaser; } });
Object.defineProperty(exports, "objectRetriever", { enumerable: true, get: function () { return transformers_1.objectRetriever; } });
Object.defineProperty(exports, "sqlite", { enumerable: true, get: function () { return transformers_1.sqlite; } });
__exportStar(require("./constants"), exports);
__exportStar(require("./entities"), exports);
__exportStar(require("./entities/types-db"), exports);
var no_xss_validator_1 = require("./utils/validators/no-xss.validator");
Object.defineProperty(exports, "NoXss", { enumerable: true, get: function () { return no_xss_validator_1.NoXss; } });
var no_url_validator_1 = require("./utils/validators/no-url.validator");
Object.defineProperty(exports, "NoUrl", { enumerable: true, get: function () { return no_url_validator_1.NoUrl; } });
__exportStar(require("./repositories"), exports);
__exportStar(require("./subscribers"), exports);
var column_1 = require("./migrations/dsl/column");
Object.defineProperty(exports, "DslColumn", { enumerable: true, get: function () { return column_1.Column; } });
var table_1 = require("./migrations/dsl/table");
Object.defineProperty(exports, "CreateTable", { enumerable: true, get: function () { return table_1.CreateTable; } });
var sqlite_1 = require("./migrations/sqlite");
Object.defineProperty(exports, "sqliteMigrations", { enumerable: true, get: function () { return sqlite_1.sqliteMigrations; } });
var mysqldb_1 = require("./migrations/mysqldb");
Object.defineProperty(exports, "mysqlMigrations", { enumerable: true, get: function () { return mysqldb_1.mysqlMigrations; } });
var postgresdb_1 = require("./migrations/postgresdb");
Object.defineProperty(exports, "postgresMigrations", { enumerable: true, get: function () { return postgresdb_1.postgresMigrations; } });
var migration_helpers_1 = require("./migrations/migration-helpers");
Object.defineProperty(exports, "wrapMigration", { enumerable: true, get: function () { return migration_helpers_1.wrapMigration; } });
__exportStar(require("./migrations/migration-types"), exports);
var db_connection_1 = require("./connection/db-connection");
Object.defineProperty(exports, "DbConnection", { enumerable: true, get: function () { return db_connection_1.DbConnection; } });
var db_connection_options_1 = require("./connection/db-connection-options");
Object.defineProperty(exports, "DbConnectionOptions", { enumerable: true, get: function () { return db_connection_options_1.DbConnectionOptions; } });
var auth_roles_service_1 = require("./services/auth.roles.service");
Object.defineProperty(exports, "AuthRolesService", { enumerable: true, get: function () { return auth_roles_service_1.AuthRolesService; } });
