// n8n-nodes-base/dist/nodes/Discord/v2/actions/message/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.sendAndWait = exports.get = exports.deleteMessage = exports.send = exports.react = exports.getAll = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const deleteMessage = __importStar(require("./deleteMessage.operation"));
exports.deleteMessage = deleteMessage;
const get = __importStar(require("./get.operation"));
exports.get = get;
const getAll = __importStar(require("./getAll.operation"));
exports.getAll = getAll;
const react = __importStar(require("./react.operation"));
exports.react = react;
const send = __importStar(require("./send.operation"));
exports.send = send;
const sendAndWait = __importStar(require("./sendAndWait.operation"));
exports.sendAndWait = sendAndWait;
const common_description_1 = require("../common.description");
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['message'],
                authentication: ['botToken', 'oAuth2'],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'deleteMessage',
                description: 'Delete a message in a channel',
                action: 'Delete a message',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Get a message in a channel',
                action: 'Get a message',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Retrieve the latest messages in a channel',
                action: 'Get many messages',
            },
            {
                name: 'React with Emoji',
                value: 'react',
                description: 'React to a message with an emoji',
                action: 'React with an emoji to a message',
            },
            {
                name: 'Send',
                value: 'send',
                description: 'Send a message to a channel, thread, or member',
                action: 'Send a message',
            },
            {
                name: 'Send and Wait for Response',
                value: n8n_workflow_1.SEND_AND_WAIT_OPERATION,
                description: 'Send a message and wait for response',
                action: 'Send message and wait for response',
            },
        ],
        default: 'send',
    },
    {
        ...common_description_1.guildRLC,
        displayOptions: {
            show: {
                resource: ['message'],
                authentication: ['botToken', 'oAuth2'],
            },
        },
    },
    ...getAll.description,
    ...react.description,
    ...send.description,
    ...deleteMessage.description,
    ...get.description,
    ...sendAndWait.description,
];
