// n8n-nodes-base/dist/nodes/Discord/v2/actions/member/roleAdd.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
exports.execute = execute;
const utilities_1 = require("../../../../../utils/utilities");
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
const common_description_1 = require("../common.description");
const properties = [common_description_1.userRLC, common_description_1.roleMultiOptions];
const displayOptions = {
    show: {
        resource: ['member'],
        operation: ['roleAdd'],
    },
    hide: {
        authentication: ['webhook'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(guildId) {
    const returnData = [];
    const items = this.getInputData();
    for (let i = 0; i < items.length; i++) {
        try {
            const userId = this.getNodeParameter('userId', i, undefined, {
                extractValue: true,
            });
            const roles = this.getNodeParameter('role', i, []);
            for (const roleId of roles) {
                await transport_1.discordApiRequest.call(this, 'PUT', `/guilds/${guildId}/members/${userId}/roles/${roleId}`);
            }
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        catch (error) {
            const err = utils_1.parseDiscordError.call(this, error, i);
            if (this.continueOnFail()) {
                returnData.push(...utils_1.prepareErrorData.call(this, err, i));
                continue;
            }
            throw err;
        }
    }
    return returnData;
}
