// @n8n/api-types/dist/dto/data-store/list-data-store-query.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListDataStoreQueryDto = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const zod_1 = require("zod");
const zod_class_1 = require("zod-class");
const pagination_dto_1 = require("../pagination/pagination.dto");
const VALID_SORT_OPTIONS = [
    'name:asc',
    'name:desc',
    'createdAt:asc',
    'createdAt:desc',
    'updatedAt:asc',
    'updatedAt:desc',
    'sizeBytes:asc',
    'sizeBytes:desc',
];
const FILTER_OPTIONS = {
    id: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    name: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
    projectId: zod_1.z.union([zod_1.z.string(), zod_1.z.array(zod_1.z.string())]).optional(),
};
const filterSchema = zod_1.z.object(FILTER_OPTIONS).strict();
const filterValidator = zod_1.z
    .string()
    .optional()
    .transform((val, ctx) => {
    if (!val)
        return undefined;
    try {
        const parsed = (0, n8n_workflow_1.jsonParse)(val);
        try {
            return filterSchema.parse(parsed);
        }
        catch (e) {
            ctx.addIssue({
                code: zod_1.z.ZodIssueCode.custom,
                message: 'Invalid filter fields',
                path: ['filter'],
            });
            return zod_1.z.NEVER;
        }
    }
    catch (e) {
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: 'Invalid filter format',
            path: ['filter'],
        });
        return zod_1.z.NEVER;
    }
});
const sortByValidator = zod_1.z
    .enum(VALID_SORT_OPTIONS, { message: `sortBy must be one of: ${VALID_SORT_OPTIONS.join(', ')}` })
    .optional();
class ListDataStoreQueryDto extends zod_class_1.Z.class({
    ...pagination_dto_1.paginationSchema,
    filter: filterValidator,
    sortBy: sortByValidator,
}) {
}
exports.ListDataStoreQueryDto = ListDataStoreQueryDto;
