// @n8n/ai-workflow-builder/dist/chains/conversation-compact.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.conversationCompactChain = conversationCompactChain;
const messages_1 = require("@langchain/core/messages");
const prompts_1 = require("@langchain/core/prompts");
const zod_1 = __importDefault(require("zod"));
const compactPromptTemplate = prompts_1.PromptTemplate.fromTemplate(`Please summarize the following conversation between a user and an AI assistant building an n8n workflow:

<previous_summary>
{previousSummary}
</previous_summary>

<conversation>
{conversationText}
</conversation>

Provide a structured summary that captures the key points, decisions made, current state of the workflow, and suggested next steps.`);
async function conversationCompactChain(llm, messages, previousSummary = '') {
    const CompactedSession = zod_1.default.object({
        summary: zod_1.default.string().describe('A concise summary of the conversation so far'),
        key_decisions: zod_1.default.array(zod_1.default.string()).describe('List of key decisions and actions taken'),
        current_state: zod_1.default.string().describe('Description of the current workflow state'),
        next_steps: zod_1.default.string().describe('Suggested next steps based on the conversation'),
    });
    const modelWithStructure = llm.withStructuredOutput(CompactedSession);
    const conversationText = messages
        .map((msg) => {
        if (msg instanceof messages_1.HumanMessage) {
            return `User: ${msg.content}`;
        }
        else if (msg instanceof messages_1.AIMessage) {
            if (typeof msg.content === 'string') {
                return `Assistant: ${msg.content}`;
            }
            else {
                return 'Assistant: Used tools';
            }
        }
        return '';
    })
        .filter(Boolean)
        .join('\n');
    const compactPrompt = await compactPromptTemplate.invoke({
        previousSummary,
        conversationText,
    });
    const structuredOutput = await modelWithStructure.invoke(compactPrompt);
    const formattedSummary = `## Previous Conversation Summary

**Summary:** ${structuredOutput.summary}

**Key Decisions:**
${structuredOutput.key_decisions.map((d) => `- ${d}`).join('\n')}

**Current State:** ${structuredOutput.current_state}

**Next Steps:** ${structuredOutput.next_steps}`;
    return {
        success: true,
        summary: structuredOutput,
        summaryPlain: formattedSummary,
    };
}
