// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/folder/update.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.properties = void 0;
exports.execute = execute;
const utilities_1 = require("../../../../../../utils/utilities");
const descriptions_1 = require("../../descriptions");
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
exports.properties = [
    descriptions_1.folderRLC,
    {
        displayName: 'Name',
        name: 'displayName',
        description: 'Name of the folder',
        type: 'string',
        default: '',
        required: true,
    },
];
const displayOptions = {
    show: {
        resource: ['folder'],
        operation: ['update'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(index) {
    const folderId = (0, utils_1.decodeOutlookId)(this.getNodeParameter('folderId', index, undefined, {
        extractValue: true,
    }));
    const displayName = this.getNodeParameter('displayName', index, undefined);
    const responseData = await transport_1.microsoftApiRequest.call(this, 'PATCH', `/mailFolders/${folderId}`, {
        displayName,
    });
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: index } });
    return executionData;
}
