// @n8n/n8n-nodes-langchain/dist/nodes/vector_store/VectorStoreZep/VectorStoreZep.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VectorStoreZep_node_exports = {};
__export(VectorStoreZep_node_exports, {
  VectorStoreZep: () => VectorStoreZep
});
module.exports = __toCommonJS(VectorStoreZep_node_exports);
var import_zep = require("@langchain/community/vectorstores/zep");
var import_zep_cloud = require("@langchain/community/vectorstores/zep_cloud");
var import_n8n_workflow = require("n8n-workflow");
var import_sharedFields = require("../../../utils/sharedFields");
var import_createVectorStoreNode = require("../shared/createVectorStoreNode/createVectorStoreNode");
const embeddingDimensions = {
  displayName: "Embedding Dimensions",
  name: "embeddingDimensions",
  type: "number",
  default: 1536,
  description: "Whether to allow using characters from the Unicode surrogate blocks"
};
const insertFields = [
  {
    displayName: "Options",
    name: "options",
    type: "collection",
    placeholder: "Add Option",
    default: {},
    options: [
      embeddingDimensions,
      {
        displayName: "Is Auto Embedded",
        name: "isAutoEmbedded",
        type: "boolean",
        default: true,
        description: "Whether to automatically embed documents when they are added"
      }
    ]
  }
];
const retrieveFields = [
  {
    displayName: "Options",
    name: "options",
    type: "collection",
    placeholder: "Add Option",
    default: {},
    options: [embeddingDimensions, import_sharedFields.metadataFilterField]
  }
];
class VectorStoreZep extends (0, import_createVectorStoreNode.createVectorStoreNode)({
  meta: {
    displayName: "Zep Vector Store",
    name: "vectorStoreZep",
    hidden: true,
    description: "Work with your data in Zep Vector Store",
    credentials: [
      {
        name: "zepApi",
        required: true
      }
    ],
    icon: "file:zep.png",
    docsUrl: "https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.vectorstorezep/"
  },
  sharedFields: [
    {
      displayName: "This Zep integration is deprecated and will be removed in a future version.",
      name: "deprecationNotice",
      type: "notice",
      default: ""
    },
    {
      displayName: "Collection Name",
      name: "collectionName",
      type: "string",
      default: "",
      required: true
    }
  ],
  insertFields,
  loadFields: retrieveFields,
  retrieveFields,
  async getVectorStoreClient(context, filter, embeddings, itemIndex) {
    const collectionName = context.getNodeParameter("collectionName", itemIndex);
    const options = context.getNodeParameter("options", itemIndex) || {};
    const credentials = await context.getCredentials("zepApi");
    const zepConfig = {
      apiKey: credentials.apiKey,
      collectionName,
      embeddingDimensions: options.embeddingDimensions ?? 1536,
      metadata: filter
    };
    if (credentials.cloud) {
      return new import_zep_cloud.ZepCloudVectorStore(embeddings, zepConfig);
    } else {
      return new import_zep.ZepVectorStore(embeddings, { ...zepConfig, apiUrl: credentials.apiUrl });
    }
  },
  async populateVectorStore(context, embeddings, documents, itemIndex) {
    const collectionName = context.getNodeParameter("collectionName", itemIndex);
    const options = context.getNodeParameter("options", itemIndex) || {};
    const credentials = await context.getCredentials("zepApi");
    const zepConfig = {
      apiKey: credentials.apiKey,
      collectionName,
      embeddingDimensions: options.embeddingDimensions ?? 1536,
      isAutoEmbedded: options.isAutoEmbedded ?? true
    };
    try {
      if (credentials.cloud) {
        await import_zep_cloud.ZepCloudVectorStore.fromDocuments(documents, embeddings, zepConfig);
      } else {
        await import_zep.ZepVectorStore.fromDocuments(documents, embeddings, {
          ...zepConfig,
          apiUrl: credentials.apiUrl
        });
      }
    } catch (error) {
      const errorCode = error.code;
      const responseData = error.responseData;
      if (errorCode === 400 && responseData.includes("CreateDocumentCollectionRequest")) {
        throw new import_n8n_workflow.NodeOperationError(context.getNode(), `Collection ${collectionName} not found`, {
          itemIndex,
          description: "Please check that the collection exists in your vector store, or make sure that collection name contains only alphanumeric characters"
        });
      }
      throw new import_n8n_workflow.NodeOperationError(context.getNode(), error, { itemIndex });
    }
  }
}) {
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VectorStoreZep
});
