// @n8n/ai-workflow-builder/dist/utils/operations-processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyOperations = applyOperations;
exports.processOperations = processOperations;
function applyOperations(workflow, operations) {
    let result = {
        nodes: [...workflow.nodes],
        connections: { ...workflow.connections },
        name: workflow.name || '',
    };
    for (const operation of operations) {
        switch (operation.type) {
            case 'clear':
                result = { nodes: [], connections: {}, name: '' };
                break;
            case 'removeNode': {
                const nodesToRemove = new Set(operation.nodeIds);
                result.nodes = result.nodes.filter((node) => !nodesToRemove.has(node.id));
                const cleanedConnections = {};
                for (const [sourceId, nodeConnections] of Object.entries(result.connections)) {
                    if (!nodesToRemove.has(sourceId)) {
                        cleanedConnections[sourceId] = {};
                        for (const [connectionType, outputs] of Object.entries(nodeConnections)) {
                            if (Array.isArray(outputs)) {
                                cleanedConnections[sourceId][connectionType] = outputs.map((outputConnections) => {
                                    if (Array.isArray(outputConnections)) {
                                        return outputConnections.filter((conn) => !nodesToRemove.has(conn.node));
                                    }
                                    return outputConnections;
                                });
                            }
                        }
                    }
                }
                result.connections = cleanedConnections;
                break;
            }
            case 'addNodes': {
                const nodeMap = new Map();
                result.nodes.forEach((node) => nodeMap.set(node.id, node));
                operation.nodes.forEach((node) => {
                    nodeMap.set(node.id, node);
                });
                result.nodes = Array.from(nodeMap.values());
                break;
            }
            case 'updateNode': {
                result.nodes = result.nodes.map((node) => {
                    if (node.id === operation.nodeId) {
                        return { ...node, ...operation.updates };
                    }
                    return node;
                });
                break;
            }
            case 'setConnections': {
                result.connections = operation.connections;
                break;
            }
            case 'mergeConnections': {
                for (const [sourceId, nodeConnections] of Object.entries(operation.connections)) {
                    if (!result.connections[sourceId]) {
                        result.connections[sourceId] = nodeConnections;
                    }
                    else {
                        for (const [connectionType, newOutputs] of Object.entries(nodeConnections)) {
                            if (!result.connections[sourceId][connectionType]) {
                                result.connections[sourceId][connectionType] = newOutputs;
                            }
                            else {
                                const existingOutputs = result.connections[sourceId][connectionType];
                                if (Array.isArray(newOutputs) && Array.isArray(existingOutputs)) {
                                    for (let i = 0; i < Math.max(newOutputs.length, existingOutputs.length); i++) {
                                        if (!newOutputs[i])
                                            continue;
                                        if (!existingOutputs[i]) {
                                            existingOutputs[i] = newOutputs[i];
                                        }
                                        else if (Array.isArray(newOutputs[i]) && Array.isArray(existingOutputs[i])) {
                                            const existingSet = new Set(existingOutputs[i].map((conn) => JSON.stringify({ node: conn.node, type: conn.type, index: conn.index })));
                                            newOutputs[i].forEach((conn) => {
                                                const connStr = JSON.stringify({
                                                    node: conn.node,
                                                    type: conn.type,
                                                    index: conn.index,
                                                });
                                                if (!existingSet.has(connStr)) {
                                                    existingOutputs[i].push(conn);
                                                }
                                            });
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                break;
            }
        }
    }
    return result;
}
function processOperations(state) {
    const { workflowJSON, workflowOperations } = state;
    if (!workflowOperations || workflowOperations.length === 0) {
        return {};
    }
    const newWorkflow = applyOperations(workflowJSON, workflowOperations);
    return {
        workflowJSON: newWorkflow,
        workflowOperations: null,
    };
}
