// n8n-nodes-base/dist/credentials/HttpHeaderAuth.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpHeaderAuth = void 0;
class HttpHeaderAuth {
    name = 'httpHeaderAuth';
    displayName = 'Header Auth';
    documentationUrl = 'httpRequest';
    genericAuth = true;
    icon = 'node:n8n-nodes-base.httpRequest';
    properties = [
        {
            displayName: 'Name',
            name: 'name',
            type: 'string',
            default: '',
        },
        {
            displayName: 'Value',
            name: 'value',
            type: 'string',
            typeOptions: {
                password: true,
            },
            default: '',
        },
        {
            displayName: 'To send multiple headers, use a "Custom Auth" credential instead',
            name: 'useCustomAuth',
            type: 'notice',
            default: '',
        },
    ];
    authenticate = {
        type: 'generic',
        properties: {
            headers: {
                '={{$credentials.name}}': '={{$credentials.value}}',
            },
        },
    };
}
exports.HttpHeaderAuth = HttpHeaderAuth;
