// n8n-nodes-base/dist/credentials/VirusTotalApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VirusTotalApi = void 0;
class VirusTotalApi {
    name = 'virusTotalApi';
    displayName = 'VirusTotal API';
    documentationUrl = 'virustotal';
    icon = 'file:icons/VirusTotal.svg';
    httpRequestNode = {
        name: 'VirusTotal',
        docsUrl: 'https://developers.virustotal.com/reference/overview',
        apiBaseUrl: 'https://www.virustotal.com/api/v3/',
    };
    properties = [
        {
            displayName: 'API Token',
            name: 'accessToken',
            type: 'string',
            typeOptions: { password: true },
            required: true,
            default: '',
        },
    ];
    authenticate = {
        type: 'generic',
        properties: {
            headers: {
                'x-apikey': '={{$credentials.accessToken}}',
            },
        },
    };
    test = {
        request: {
            baseURL: 'https://www.virustotal.com/api/v3',
            url: '/popular_threat_categories',
        },
    };
}
exports.VirusTotalApi = VirusTotalApi;
