// @n8n/db/dist/migrations/common/1750252139168-LinkRoleToUserTable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkRoleToUserTable1750252139168 = void 0;
class LinkRoleToUserTable1750252139168 {
    async up({ schemaBuilder: { addForeignKey, addColumns, column }, escape, dbType, runQuery, }) {
        const roleTableName = escape.tableName('role');
        const userTableName = escape.tableName('user');
        const slugColumn = escape.columnName('slug');
        const roleColumn = escape.columnName('role');
        const roleSlugColumn = escape.columnName('roleSlug');
        const roleTypeColumn = escape.columnName('roleType');
        const systemRoleColumn = escape.columnName('systemRole');
        const isPostgresOrSqlite = dbType === 'postgresdb' || dbType === 'sqlite';
        const upsertQuery = isPostgresOrSqlite
            ? `INSERT INTO ${roleTableName} (${slugColumn}, ${roleTypeColumn}, ${systemRoleColumn}) VALUES (:slug, :roleType, :systemRole) ON CONFLICT DO NOTHING`
            : `INSERT IGNORE INTO ${roleTableName} (${slugColumn}, ${roleTypeColumn}, ${systemRoleColumn}) VALUES (:slug, :roleType, :systemRole)`;
        for (const role of ['global:owner', 'global:admin', 'global:member']) {
            await runQuery(upsertQuery, {
                slug: role,
                roleType: 'global',
                systemRole: true,
            });
        }
        await addColumns('user', [column('roleSlug').varchar(128).default("'global:member'").notNull]);
        await runQuery(`UPDATE ${userTableName} SET ${roleSlugColumn} = ${roleColumn} WHERE ${roleColumn} != ${roleSlugColumn}`);
        await runQuery(`UPDATE ${userTableName} SET ${roleSlugColumn} = 'global:member' WHERE NOT EXISTS (SELECT 1 FROM ${roleTableName} WHERE ${slugColumn} = ${roleSlugColumn})`);
        await addForeignKey('user', 'roleSlug', ['role', 'slug']);
    }
    async down({ schemaBuilder: { dropForeignKey, dropColumns } }) {
        await dropForeignKey('user', 'roleSlug', ['role', 'slug']);
        await dropColumns('user', ['roleSlug']);
    }
}
exports.LinkRoleToUserTable1750252139168 = LinkRoleToUserTable1750252139168;
