// @n8n/task-runner/dist/js-task-runner/errors/execution-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecutionError = void 0;
const serializable_error_1 = require("./serializable-error");
class ExecutionError extends serializable_error_1.SerializableError {
    constructor(error, itemIndex) {
        super(error.message);
        this.description = null;
        this.itemIndex = undefined;
        this.context = undefined;
        this.lineNumber = undefined;
        this.itemIndex = itemIndex;
        if (this.itemIndex !== undefined) {
            this.context = { itemIndex: this.itemIndex };
        }
        Object.defineProperty(this, 'stack', {
            value: error.stack,
            enumerable: true,
        });
        this.populateFromStack();
    }
    populateFromStack() {
        const stackRows = (this.stack ?? '').split('\n');
        if (stackRows.length === 0) {
            this.message = 'Unknown error';
            return;
        }
        const messageRow = stackRows.find((line) => line.includes('Error:'));
        const lineNumberDisplay = this.toLineNumberDisplay(stackRows);
        if (!messageRow) {
            this.message = `Unknown error ${lineNumberDisplay}`;
            return;
        }
        const [errorDetails, errorType] = this.toErrorDetailsAndType(messageRow);
        if (errorType)
            this.description = errorType;
        if (!errorDetails) {
            this.message = `Unknown error ${lineNumberDisplay}`;
            return;
        }
        this.message = `${errorDetails} ${lineNumberDisplay}`;
    }
    toLineNumberDisplay(stackRows) {
        if (!stackRows || stackRows.length === 0)
            return '';
        const userFnLine = stackRows.find((row) => row.match(/\(evalmachine\.<anonymous>:\d+:\d+\)/) && !row.includes('VmCodeWrapper'));
        if (userFnLine) {
            const match = userFnLine.match(/evalmachine\.<anonymous>:(\d+):/);
            if (match)
                this.lineNumber = Number(match[1]);
        }
        if (this.lineNumber === undefined) {
            const topLevelLine = stackRows.find((row) => row.includes('VmCodeWrapper') && row.includes('evalmachine.<anonymous>'));
            if (topLevelLine) {
                const match = topLevelLine.match(/evalmachine\.<anonymous>:(\d+):/);
                if (match)
                    this.lineNumber = Number(match[1]);
            }
        }
        if (this.lineNumber === undefined)
            return '';
        return this.itemIndex === undefined
            ? `[line ${this.lineNumber}]`
            : `[line ${this.lineNumber}, for item ${this.itemIndex}]`;
    }
    toErrorDetailsAndType(messageRow) {
        if (!messageRow)
            return [null, null];
        const segments = messageRow.split(':').map((i) => i.trim());
        if (segments[1] === "Cannot find module 'node") {
            segments[1] = `${segments[1]}:${segments[2]}`;
            segments.splice(2, 1);
        }
        if (segments.length >= 3 &&
            segments[1]?.startsWith("Module 'node") &&
            segments[2]?.includes("' is disallowed")) {
            segments[1] = `${segments[1]}:${segments[2]}`;
            segments.splice(2, 1);
        }
        const [errorDetails, errorType] = segments.reverse();
        return [errorDetails, errorType === 'Error' ? null : errorType];
    }
}
exports.ExecutionError = ExecutionError;
