// n8n-nodes-base/dist/nodes/Splunk/v2/helpers/descriptions/rlc.description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userRLC = exports.searchJobRLC = exports.reportRLC = void 0;
exports.reportRLC = {
    displayName: 'Report',
    name: 'reportId',
    type: 'resourceLocator',
    default: { mode: 'list', value: '' },
    required: true,
    modes: [
        {
            displayName: 'From List',
            name: 'list',
            type: 'list',
            placeholder: 'Select a report...',
            typeOptions: {
                searchListMethod: 'searchReports',
                searchable: true,
            },
        },
        {
            displayName: 'ID',
            name: 'id',
            type: 'string',
            placeholder: 'e.g. Errors%20in%20the%20last%20hour',
        },
    ],
};
exports.searchJobRLC = {
    displayName: 'Search Job',
    name: 'searchJobId',
    type: 'resourceLocator',
    default: { mode: 'list', value: '' },
    required: true,
    modes: [
        {
            displayName: 'From List',
            name: 'list',
            type: 'list',
            placeholder: 'Select a search job...',
            typeOptions: {
                searchListMethod: 'searchJobs',
                searchable: true,
            },
        },
        {
            displayName: 'ID',
            name: 'id',
            type: 'string',
            placeholder: 'e.g. 1718944376.178',
        },
    ],
};
exports.userRLC = {
    displayName: 'User',
    name: 'userId',
    type: 'resourceLocator',
    default: { mode: 'list', value: '' },
    required: true,
    modes: [
        {
            displayName: 'From List',
            name: 'list',
            type: 'list',
            placeholder: 'Select a user...',
            typeOptions: {
                searchListMethod: 'searchUsers',
                searchable: true,
            },
        },
        {
            displayName: 'ID',
            name: 'id',
            type: 'string',
            placeholder: 'e.g. admin',
        },
    ],
};
