// @n8n/db/dist/entities/role.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Role = void 0;
const typeorm_1 = require("@n8n/typeorm");
const abstract_entity_1 = require("./abstract-entity");
const scope_1 = require("./scope");
let Role = class Role extends abstract_entity_1.WithTimestamps {
};
exports.Role = Role;
__decorate([
    (0, typeorm_1.PrimaryColumn)({
        type: String,
        name: 'slug',
    }),
    __metadata("design:type", String)
], Role.prototype, "slug", void 0);
__decorate([
    (0, typeorm_1.Column)({
        type: String,
        nullable: false,
        name: 'displayName',
    }),
    __metadata("design:type", String)
], Role.prototype, "displayName", void 0);
__decorate([
    (0, typeorm_1.Column)({
        type: String,
        nullable: true,
        name: 'description',
    }),
    __metadata("design:type", Object)
], Role.prototype, "description", void 0);
__decorate([
    (0, typeorm_1.Column)({
        type: Boolean,
        default: false,
        name: 'systemRole',
    }),
    __metadata("design:type", Boolean)
], Role.prototype, "systemRole", void 0);
__decorate([
    (0, typeorm_1.Column)({
        type: String,
        name: 'roleType',
    }),
    __metadata("design:type", String)
], Role.prototype, "roleType", void 0);
__decorate([
    (0, typeorm_1.OneToMany)('ProjectRelation', 'role'),
    __metadata("design:type", Array)
], Role.prototype, "projectRelations", void 0);
__decorate([
    (0, typeorm_1.ManyToMany)(() => scope_1.Scope, {
        eager: true,
    }),
    (0, typeorm_1.JoinTable)({
        name: 'role_scope',
        joinColumn: { name: 'roleSlug', referencedColumnName: 'slug' },
        inverseJoinColumn: { name: 'scopeSlug', referencedColumnName: 'slug' },
    }),
    __metadata("design:type", Array)
], Role.prototype, "scopes", void 0);
exports.Role = Role = __decorate([
    (0, typeorm_1.Entity)({
        name: 'role',
    })
], Role);
