// @n8n/n8n-nodes-langchain/dist/nodes/llms/LmChatVercelAiGateway/LmChatVercelAiGateway.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LmChatVercelAiGateway_node_exports = {};
__export(LmChatVercelAiGateway_node_exports, {
  LmChatVercelAiGateway: () => LmChatVercelAiGateway
});
module.exports = __toCommonJS(LmChatVercelAiGateway_node_exports);
var import_openai = require("@langchain/openai");
var import_n8n_workflow = require("n8n-workflow");
var import_httpProxyAgent = require("../../../utils/httpProxyAgent");
var import_sharedFields = require("../../../utils/sharedFields");
var import_error_handling = require("../../vendors/OpenAi/helpers/error-handling");
var import_n8nLlmFailedAttemptHandler = require("../n8nLlmFailedAttemptHandler");
var import_N8nLlmTracing = require("../N8nLlmTracing");
class LmChatVercelAiGateway {
  constructor() {
    this.description = {
      displayName: "Vercel AI Gateway Chat Model",
      name: "lmChatVercelAiGateway",
      icon: { light: "file:vercel.dark.svg", dark: "file:vercel.svg" },
      group: ["transform"],
      version: [1],
      description: "For advanced usage with an AI chain via Vercel AI Gateway",
      defaults: {
        name: "Vercel AI Gateway Chat Model"
      },
      codex: {
        categories: ["AI"],
        subcategories: {
          AI: ["Language Models", "Root Nodes"],
          "Language Models": ["Chat Models (Recommended)"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.lmchatvercel/"
            }
          ]
        }
      },
      inputs: [],
      outputs: [import_n8n_workflow.NodeConnectionTypes.AiLanguageModel],
      outputNames: ["Model"],
      credentials: [
        {
          name: "vercelAiGatewayApi",
          required: true
        }
      ],
      requestDefaults: {
        ignoreHttpStatusErrors: true,
        baseURL: "={{ $credentials?.url }}"
      },
      properties: [
        (0, import_sharedFields.getConnectionHintNoticeField)([import_n8n_workflow.NodeConnectionTypes.AiChain, import_n8n_workflow.NodeConnectionTypes.AiAgent]),
        {
          displayName: 'If using JSON response format, you must include word "json" in the prompt in your chain or agent. Also, make sure to select latest models released post November 2023.',
          name: "notice",
          type: "notice",
          default: "",
          displayOptions: {
            show: {
              "/options.responseFormat": ["json_object"]
            }
          }
        },
        {
          displayName: "Model",
          name: "model",
          type: "options",
          description: "The model which will generate the completion",
          typeOptions: {
            loadOptions: {
              routing: {
                request: {
                  method: "GET",
                  url: "/models"
                },
                output: {
                  postReceive: [
                    {
                      type: "rootProperty",
                      properties: {
                        property: "data"
                      }
                    },
                    {
                      type: "setKeyValue",
                      properties: {
                        name: "={{$responseItem.id}}",
                        value: "={{$responseItem.id}}"
                      }
                    },
                    {
                      type: "sort",
                      properties: {
                        key: "name"
                      }
                    }
                  ]
                }
              }
            }
          },
          routing: {
            send: {
              type: "body",
              property: "model"
            }
          },
          default: "openai/gpt-4o"
        },
        {
          displayName: "Options",
          name: "options",
          placeholder: "Add Option",
          description: "Additional options to add",
          type: "collection",
          default: {},
          options: [
            {
              displayName: "Frequency Penalty",
              name: "frequencyPenalty",
              default: 0,
              typeOptions: { maxValue: 2, minValue: -2, numberPrecision: 1 },
              description: "Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim",
              type: "number"
            },
            {
              displayName: "Maximum Number of Tokens",
              name: "maxTokens",
              default: -1,
              description: "The maximum number of tokens to generate in the completion. Most models have a context length of 2048 tokens (except for the newest models, which support 32,768).",
              type: "number",
              typeOptions: {
                maxValue: 32768
              }
            },
            {
              displayName: "Response Format",
              name: "responseFormat",
              default: "text",
              type: "options",
              options: [
                {
                  name: "Text",
                  value: "text",
                  description: "Regular text response"
                },
                {
                  name: "JSON",
                  value: "json_object",
                  description: "Enables JSON mode, which should guarantee the message the model generates is valid JSON"
                }
              ]
            },
            {
              displayName: "Presence Penalty",
              name: "presencePenalty",
              default: 0,
              typeOptions: { maxValue: 2, minValue: -2, numberPrecision: 1 },
              description: "Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics",
              type: "number"
            },
            {
              displayName: "Sampling Temperature",
              name: "temperature",
              default: 0.7,
              typeOptions: { maxValue: 2, minValue: 0, numberPrecision: 1 },
              description: "Controls randomness: Lowering results in less random completions. As the temperature approaches zero, the model will become deterministic and repetitive.",
              type: "number"
            },
            {
              displayName: "Timeout",
              name: "timeout",
              default: 36e4,
              description: "Maximum amount of time a request is allowed to take in milliseconds",
              type: "number"
            },
            {
              displayName: "Max Retries",
              name: "maxRetries",
              default: 2,
              description: "Maximum number of retries to attempt",
              type: "number"
            },
            {
              displayName: "Top P",
              name: "topP",
              default: 1,
              typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
              description: "Controls diversity via nucleus sampling: 0.5 means half of all likelihood-weighted options are considered. We generally recommend altering this or temperature but not both.",
              type: "number"
            }
          ]
        }
      ]
    };
  }
  async supplyData(itemIndex) {
    const credentials = await this.getCredentials("vercelAiGatewayApi");
    const modelName = this.getNodeParameter("model", itemIndex);
    const options = this.getNodeParameter("options", itemIndex, {});
    const configuration = {
      baseURL: credentials.url,
      fetchOptions: {
        dispatcher: (0, import_httpProxyAgent.getProxyAgent)(credentials.url)
      }
    };
    const model = new import_openai.ChatOpenAI({
      apiKey: credentials.apiKey,
      model: modelName,
      ...options,
      timeout: options.timeout ?? 6e4,
      maxRetries: options.maxRetries ?? 2,
      configuration,
      callbacks: [new import_N8nLlmTracing.N8nLlmTracing(this)],
      modelKwargs: options.responseFormat ? {
        response_format: { type: options.responseFormat }
      } : void 0,
      onFailedAttempt: (0, import_n8nLlmFailedAttemptHandler.makeN8nLlmFailedAttemptHandler)(this, import_error_handling.openAiFailedAttemptHandler)
    });
    return {
      response: model
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LmChatVercelAiGateway
});
