// @n8n/backend-test-utils/dist/test-db.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBootstrapDBOptions = exports.testDbPrefix = void 0;
exports.init = init;
exports.isReady = isReady;
exports.terminate = terminate;
exports.truncate = truncate;
const config_1 = require("@n8n/config");
const db_1 = require("@n8n/db");
const di_1 = require("@n8n/di");
const typeorm_1 = require("@n8n/typeorm");
const n8n_workflow_1 = require("n8n-workflow");
exports.testDbPrefix = 'n8n_test_';
let isInitialized = false;
const getBootstrapDBOptions = (dbType) => {
    const globalConfig = di_1.Container.get(config_1.GlobalConfig);
    const type = dbType === 'postgresdb' ? 'postgres' : 'mysql';
    return {
        type,
        ...di_1.Container.get(db_1.DbConnectionOptions).getOverrides(dbType),
        database: type,
        entityPrefix: globalConfig.database.tablePrefix,
        schema: dbType === 'postgresdb' ? globalConfig.database.postgresdb.schema : undefined,
    };
};
exports.getBootstrapDBOptions = getBootstrapDBOptions;
async function init() {
    if (isInitialized)
        return;
    const globalConfig = di_1.Container.get(config_1.GlobalConfig);
    const dbType = globalConfig.database.type;
    const testDbName = `${exports.testDbPrefix}${(0, n8n_workflow_1.randomString)(6, 10).toLowerCase()}_${Date.now()}`;
    if (dbType === 'postgresdb') {
        const bootstrapPostgres = await new typeorm_1.DataSource((0, exports.getBootstrapDBOptions)('postgresdb')).initialize();
        await bootstrapPostgres.query(`CREATE DATABASE ${testDbName}`);
        await bootstrapPostgres.destroy();
        globalConfig.database.postgresdb.database = testDbName;
    }
    else if (dbType === 'mysqldb' || dbType === 'mariadb') {
        const bootstrapMysql = await new typeorm_1.DataSource((0, exports.getBootstrapDBOptions)('mysqldb')).initialize();
        await bootstrapMysql.query(`CREATE DATABASE ${testDbName} DEFAULT CHARACTER SET utf8mb4`);
        await bootstrapMysql.destroy();
        globalConfig.database.mysqldb.database = testDbName;
    }
    const dbConnection = di_1.Container.get(db_1.DbConnection);
    await dbConnection.init();
    await dbConnection.migrate();
    await di_1.Container.get(db_1.AuthRolesService).init();
    isInitialized = true;
}
function isReady() {
    const { connectionState } = di_1.Container.get(db_1.DbConnection);
    return connectionState.connected && connectionState.migrated;
}
async function terminate() {
    const dbConnection = di_1.Container.get(db_1.DbConnection);
    await dbConnection.close();
    dbConnection.connectionState.connected = false;
    isInitialized = false;
}
async function truncate(entities) {
    const connection = di_1.Container.get(typeorm_1.DataSource);
    for (const name of entities) {
        await connection.getRepository(name).delete({});
    }
}
