// @n8n/api-types/dist/schemas/data-table-filter.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataTableFilterSchema = exports.dataTableFilterTypeSchema = exports.dataTableFilterRecordSchema = exports.FilterConditionSchema = void 0;
const zod_1 = require("zod");
const data_store_schema_1 = require("./data-store.schema");
exports.FilterConditionSchema = zod_1.z.union([
    zod_1.z.literal('eq'),
    zod_1.z.literal('neq'),
    zod_1.z.literal('like'),
    zod_1.z.literal('ilike'),
    zod_1.z.literal('gt'),
    zod_1.z.literal('gte'),
    zod_1.z.literal('lt'),
    zod_1.z.literal('lte'),
]);
exports.dataTableFilterRecordSchema = zod_1.z.object({
    columnName: data_store_schema_1.dataStoreColumnNameSchema,
    condition: exports.FilterConditionSchema.default('eq'),
    value: zod_1.z.union([zod_1.z.string(), zod_1.z.number(), zod_1.z.boolean(), zod_1.z.date(), zod_1.z.null()]),
});
exports.dataTableFilterTypeSchema = zod_1.z.union([zod_1.z.literal('and'), zod_1.z.literal('or')]);
exports.dataTableFilterSchema = zod_1.z.object({
    type: exports.dataTableFilterTypeSchema.default('and'),
    filters: zod_1.z.array(exports.dataTableFilterRecordSchema).default([]),
});
