// @n8n/api-types/dist/schemas/user.schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.usersListSchema = exports.userListItemSchema = exports.userProjectSchema = exports.roleSchema = exports.ROLE = void 0;
const permissions_1 = require("@n8n/permissions");
const zod_1 = require("zod");
const user_settings_schema_1 = require("./user-settings.schema");
exports.ROLE = {
    Owner: 'global:owner',
    Member: 'global:member',
    Admin: 'global:admin',
    Default: 'default',
};
const roleValuesForSchema = Object.values(exports.ROLE);
exports.roleSchema = zod_1.z.enum(roleValuesForSchema);
exports.userProjectSchema = zod_1.z.object({
    id: zod_1.z.string(),
    role: permissions_1.projectRoleSchema,
    name: zod_1.z.string(),
});
exports.userListItemSchema = zod_1.z.object({
    id: zod_1.z.string(),
    firstName: zod_1.z.string().nullable().optional(),
    lastName: zod_1.z.string().nullable().optional(),
    email: zod_1.z.string().email().nullable().optional(),
    role: exports.roleSchema.optional(),
    isPending: zod_1.z.boolean().optional(),
    isOwner: zod_1.z.boolean().optional(),
    signInType: zod_1.z.string().optional(),
    settings: user_settings_schema_1.userSettingsSchema.nullable().optional(),
    personalizationAnswers: zod_1.z.object({}).passthrough().nullable().optional(),
    projectRelations: zod_1.z.array(exports.userProjectSchema).nullable().optional(),
    mfaEnabled: zod_1.z.boolean().optional(),
    lastActiveAt: zod_1.z.string().nullable().optional(),
    inviteAcceptUrl: zod_1.z.string().optional(),
});
exports.usersListSchema = zod_1.z.object({
    count: zod_1.z.number(),
    items: zod_1.z.array(exports.userListItemSchema),
});
