// @n8n/n8n-nodes-langchain/dist/credentials/AzureOpenAiApi.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AzureOpenAiApi_credentials_exports = {};
__export(AzureOpenAiApi_credentials_exports, {
  AzureOpenAiApi: () => AzureOpenAiApi
});
module.exports = __toCommonJS(AzureOpenAiApi_credentials_exports);
class AzureOpenAiApi {
  constructor() {
    this.name = "azureOpenAiApi";
    this.displayName = "Azure Open AI";
    this.documentationUrl = "azureopenai";
    this.properties = [
      {
        displayName: "API Key",
        name: "apiKey",
        type: "string",
        typeOptions: { password: true },
        required: true,
        default: ""
      },
      {
        displayName: "Resource Name",
        name: "resourceName",
        type: "string",
        required: true,
        default: ""
      },
      {
        displayName: "API Version",
        name: "apiVersion",
        type: "string",
        required: true,
        default: "2025-03-01-preview"
      },
      {
        displayName: "Endpoint",
        name: "endpoint",
        type: "string",
        default: void 0,
        placeholder: "https://westeurope.api.cognitive.microsoft.com"
      }
    ];
    this.authenticate = {
      type: "generic",
      properties: {
        headers: {
          "api-key": "={{$credentials.apiKey}}"
        }
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureOpenAiApi
});
