// @n8n/n8n-nodes-langchain/dist/credentials/ZepApi.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var ZepApi_credentials_exports = {};
__export(ZepApi_credentials_exports, {
  ZepApi: () => ZepApi
});
module.exports = __toCommonJS(ZepApi_credentials_exports);
class ZepApi {
  constructor() {
    this.name = "zepApi";
    this.displayName = "Zep Api";
    this.documentationUrl = "zep";
    this.properties = [
      {
        displayName: "This Zep integration is deprecated and will be removed in a future version.",
        name: "deprecationNotice",
        type: "notice",
        default: ""
      },
      {
        displayName: "API Key",
        name: "apiKey",
        type: "string",
        typeOptions: { password: true },
        required: false,
        default: ""
      },
      {
        displayName: "Cloud",
        description: "Whether you are adding credentials for Zep Cloud instead of Zep Open Source",
        name: "cloud",
        type: "boolean",
        default: false
      },
      {
        displayName: "API URL",
        name: "apiUrl",
        required: false,
        type: "string",
        default: "http://localhost:8000",
        displayOptions: {
          show: {
            cloud: [false]
          }
        }
      }
    ];
    this.authenticate = {
      type: "generic",
      properties: {
        headers: {
          Authorization: '={{$credentials.apiKey && !$credentials.cloud ? "Bearer " + $credentials.apiKey : "Api-Key " + $credentials.apiKey }}'
        }
      }
    };
    this.test = {
      request: {
        baseURL: '={{!$credentials.cloud ? $credentials.apiUrl : "https://api.getzep.com"}}',
        url: '={{!$credentials.cloud ? "/api/v1/collection" : "/api/v2/collections"}}'
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ZepApi
});
