// n8n-nodes-base/dist/nodes/TheHiveProject/actions/alert/merge.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
exports.execute = execute;
const utilities_1 = require("../../../../utils/utilities");
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const properties = [descriptions_1.alertRLC, descriptions_1.caseRLC];
const displayOptions = {
    show: {
        resource: ['alert'],
        operation: ['merge'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const alertId = this.getNodeParameter('alertId', i, '', { extractValue: true });
    const caseId = this.getNodeParameter('caseId', i, '', { extractValue: true });
    responseData = await transport_1.theHiveApiRequest.call(this, 'POST', `/alert/${alertId}/merge/${caseId}`, {});
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
