// @n8n/ai-workflow-builder/dist/tools/utils/connection.utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateConnection = validateConnection;
exports.nodeHasOutputType = nodeHasOutputType;
exports.nodeAcceptsInputType = nodeAcceptsInputType;
exports.createConnection = createConnection;
exports.removeConnection = removeConnection;
exports.getNodeConnections = getNodeConnections;
exports.formatConnectionMessage = formatConnectionMessage;
exports.inferConnectionType = inferConnectionType;
const n8n_workflow_1 = require("n8n-workflow");
const node_helpers_1 = require("../../utils/node-helpers");
function extractConnectionTypesFromExpression(expression) {
    const types = new Set();
    const patterns = [/type\s*:\s*["']([^"']+)["']/g, /type\s*:\s*NodeConnectionTypes\.(\w+)/g];
    const arrayMainPatterns = [
        /\[\s*["']main["']/i,
        /\[\s*NodeConnectionTypes\.Main/i,
        /return\s+\[\s*["']main["']/i,
        /return\s+\[\s*NodeConnectionTypes\.Main/i,
    ];
    for (const pattern of arrayMainPatterns) {
        if (pattern.test(expression)) {
            types.add(n8n_workflow_1.NodeConnectionTypes.Main);
            break;
        }
    }
    for (const pattern of patterns) {
        let match;
        pattern.lastIndex = 0;
        while ((match = pattern.exec(expression)) !== null) {
            const type = match[1];
            if (type) {
                const normalizedType = type.toLowerCase() === 'main' ? n8n_workflow_1.NodeConnectionTypes.Main : type;
                types.add(normalizedType);
            }
        }
    }
    return Array.from(types);
}
function validateConnection(sourceNode, targetNode, connectionType, nodeTypes) {
    const sourceNodeType = nodeTypes.find((nt) => nt.name === sourceNode.type);
    const targetNodeType = nodeTypes.find((nt) => nt.name === targetNode.type);
    if (!sourceNodeType || !targetNodeType) {
        return {
            valid: false,
            error: 'One or both node types not found',
        };
    }
    const sourceIsSubNode = (0, node_helpers_1.isSubNode)(sourceNodeType, sourceNode);
    const targetIsSubNode = (0, node_helpers_1.isSubNode)(targetNodeType, targetNode);
    if (connectionType.startsWith('ai_')) {
        if (!sourceIsSubNode && !targetIsSubNode) {
            return {
                valid: false,
                error: `Connection type "${connectionType}" requires a sub-node, but both nodes are main nodes`,
            };
        }
        if (targetIsSubNode && !sourceIsSubNode) {
            return {
                valid: true,
                shouldSwap: true,
                swappedSource: targetNode,
                swappedTarget: sourceNode,
            };
        }
        if (sourceIsSubNode) {
            const supportsConnectionType = nodeHasOutputType(sourceNodeType, connectionType);
            if (!supportsConnectionType) {
                return {
                    valid: false,
                    error: `Sub-node "${sourceNode.name}" does not support output type "${connectionType}"`,
                };
            }
        }
    }
    return { valid: true };
}
function nodeHasOutputType(nodeType, connectionType) {
    if (typeof nodeType.outputs === 'string') {
        return nodeType.outputs === connectionType || nodeType.outputs.includes(connectionType);
    }
    if (!nodeType.outputs || !Array.isArray(nodeType.outputs)) {
        return false;
    }
    return nodeType.outputs.some((output) => {
        if (typeof output === 'string') {
            return output === connectionType || output.includes(connectionType);
        }
        return output.type === connectionType;
    });
}
function nodeAcceptsInputType(nodeType, connectionType) {
    if (typeof nodeType.inputs === 'string') {
        return nodeType.inputs === connectionType || nodeType.inputs.includes(connectionType);
    }
    if (!nodeType.inputs || !Array.isArray(nodeType.inputs)) {
        return false;
    }
    return nodeType.inputs.some((input) => {
        if (typeof input === 'string') {
            return input === connectionType || input.includes(connectionType);
        }
        return input.type === connectionType;
    });
}
function createConnection(connections, sourceNodeName, targetNodeName, connectionType, sourceOutputIndex = 0, targetInputIndex = 0) {
    if (!connections[sourceNodeName]) {
        connections[sourceNodeName] = {};
    }
    if (!connections[sourceNodeName][connectionType]) {
        connections[sourceNodeName][connectionType] = [];
    }
    const connectionArray = connections[sourceNodeName][connectionType];
    while (connectionArray.length <= sourceOutputIndex) {
        connectionArray.push([]);
    }
    const newConnection = {
        node: targetNodeName,
        type: connectionType,
        index: targetInputIndex,
    };
    connectionArray[sourceOutputIndex] ??= [];
    const existingConnection = connectionArray[sourceOutputIndex].find((conn) => conn.node === targetNodeName && conn.index === targetInputIndex);
    if (!existingConnection) {
        connectionArray[sourceOutputIndex].push(newConnection);
    }
    return connections;
}
function removeConnection(connections, sourceNodeName, targetNodeName, connectionType, sourceOutputIndex, targetInputIndex) {
    if (!connections[sourceNodeName]?.[connectionType]) {
        return connections;
    }
    const connectionArray = connections[sourceNodeName][connectionType];
    if (sourceOutputIndex !== undefined) {
        if (connectionArray[sourceOutputIndex]) {
            connectionArray[sourceOutputIndex] = connectionArray[sourceOutputIndex].filter((conn) => conn.node !== targetNodeName ||
                (targetInputIndex !== undefined && conn.index !== targetInputIndex));
        }
    }
    else {
        for (let i = 0; i < connectionArray.length; i++) {
            if (connectionArray[i]) {
                connectionArray[i] = connectionArray[i].filter((conn) => conn.node !== targetNodeName);
            }
        }
    }
    connections[sourceNodeName][connectionType] = connectionArray.filter((arr) => arr && arr.length > 0);
    if (connections[sourceNodeName][connectionType].length === 0) {
        delete connections[sourceNodeName][connectionType];
    }
    if (Object.keys(connections[sourceNodeName]).length === 0) {
        delete connections[sourceNodeName];
    }
    return connections;
}
function getNodeConnections(connections, nodeName, direction) {
    const result = [];
    if (direction === 'source') {
        const nodeConnections = connections[nodeName];
        if (nodeConnections) {
            for (const [connectionType, connectionArray] of Object.entries(nodeConnections)) {
                connectionArray.forEach((outputConnections, sourceIndex) => {
                    if (outputConnections) {
                        outputConnections.forEach((conn) => {
                            result.push({
                                node: conn.node,
                                type: connectionType,
                                sourceIndex,
                                targetIndex: conn.index,
                            });
                        });
                    }
                });
            }
        }
    }
    else {
        for (const [sourceNode, nodeConnections] of Object.entries(connections)) {
            for (const [connectionType, connectionArray] of Object.entries(nodeConnections)) {
                connectionArray.forEach((outputConnections, sourceIndex) => {
                    if (outputConnections) {
                        outputConnections.forEach((conn) => {
                            if (conn.node === nodeName) {
                                result.push({
                                    node: sourceNode,
                                    type: connectionType,
                                    sourceIndex,
                                    targetIndex: conn.index,
                                });
                            }
                        });
                    }
                });
            }
        }
    }
    return result;
}
function formatConnectionMessage(sourceNode, targetNode, connectionType, swapped = false) {
    if (swapped) {
        return `Auto-corrected connection: ${sourceNode} (${connectionType}) → ${targetNode}. (Note: Swapped nodes to ensure sub-node is the source)`;
    }
    return `Connected: ${sourceNode} → ${targetNode} (${connectionType})`;
}
function getNodeOutputTypes(nodeType) {
    if (typeof nodeType.outputs === 'string') {
        const extracted = extractConnectionTypesFromExpression(nodeType.outputs);
        if (extracted.length > 0) {
            return extracted;
        }
        return [];
    }
    if (!nodeType.outputs || !Array.isArray(nodeType.outputs)) {
        return [];
    }
    return nodeType.outputs.map((output) => {
        if (typeof output === 'string') {
            return output;
        }
        return output.type;
    });
}
function getNodeInputTypes(nodeType, node) {
    if (typeof nodeType.inputs === 'string') {
        if (node &&
            nodeType.name.includes('vectorStore') &&
            node.parameters?.mode === 'retrieve-as-tool') {
            const extracted = extractConnectionTypesFromExpression(nodeType.inputs);
            return extracted.filter((type) => type.startsWith('ai_'));
        }
        const extracted = extractConnectionTypesFromExpression(nodeType.inputs);
        if (extracted.length > 0) {
            return extracted;
        }
        return [];
    }
    if (!nodeType.inputs || !Array.isArray(nodeType.inputs)) {
        return [];
    }
    return nodeType.inputs.map((input) => {
        if (typeof input === 'string') {
            return input;
        }
        return input.type;
    });
}
function inferConnectionType(sourceNode, targetNode, sourceNodeType, targetNodeType) {
    const sourceOutputTypes = getNodeOutputTypes(sourceNodeType);
    const targetInputTypes = getNodeInputTypes(targetNodeType, targetNode);
    const sourceInputTypes = getNodeInputTypes(sourceNodeType, sourceNode);
    const sourceHasMainInput = sourceInputTypes.includes(n8n_workflow_1.NodeConnectionTypes.Main);
    const targetHasMainInput = targetInputTypes.includes(n8n_workflow_1.NodeConnectionTypes.Main);
    const sourceIsSubNode = (0, node_helpers_1.isSubNode)(sourceNodeType, sourceNode) ||
        (typeof sourceNodeType.inputs === 'string' && !sourceHasMainInput);
    const targetIsSubNode = (0, node_helpers_1.isSubNode)(targetNodeType, targetNode) ||
        (typeof targetNodeType.inputs === 'string' && !targetHasMainInput);
    const matchingTypes = sourceOutputTypes.filter((outputType) => targetInputTypes.includes(outputType));
    if (sourceIsSubNode && !targetIsSubNode) {
        const aiConnectionTypes = matchingTypes.filter((type) => type.startsWith('ai_'));
        if (aiConnectionTypes.length === 1) {
            return { connectionType: aiConnectionTypes[0] };
        }
        else if (aiConnectionTypes.length > 1) {
            return {
                possibleTypes: aiConnectionTypes,
                error: `Multiple AI connection types possible: ${aiConnectionTypes.join(', ')}. Please specify which one to use.`,
            };
        }
    }
    if (!sourceIsSubNode && targetIsSubNode) {
        const targetOutputTypes = getNodeOutputTypes(targetNodeType);
        const sourceInputTypes = getNodeInputTypes(sourceNodeType, sourceNode);
        const reverseAiMatches = targetOutputTypes
            .filter((type) => type.startsWith('ai_'))
            .filter((type) => sourceInputTypes.includes(type));
        if (reverseAiMatches.length === 1) {
            return {
                connectionType: reverseAiMatches[0],
                requiresSwap: true,
            };
        }
        else if (reverseAiMatches.length > 1) {
            return {
                possibleTypes: reverseAiMatches,
                requiresSwap: true,
                error: `Multiple AI connection types possible (requires swap): ${reverseAiMatches.join(', ')}. Please specify which one to use.`,
            };
        }
    }
    if (!sourceIsSubNode && !targetIsSubNode) {
        if (matchingTypes.includes(n8n_workflow_1.NodeConnectionTypes.Main)) {
            return { connectionType: n8n_workflow_1.NodeConnectionTypes.Main };
        }
    }
    if (sourceIsSubNode && targetIsSubNode) {
        const subNodeConnections = matchingTypes.filter((type) => type.startsWith('ai_'));
        if (subNodeConnections.length === 1) {
            return { connectionType: subNodeConnections[0] };
        }
        else if (subNodeConnections.length > 1) {
            return {
                possibleTypes: subNodeConnections,
                error: `Multiple connection types possible between sub-nodes: ${subNodeConnections.join(', ')}. Please specify which one to use.`,
            };
        }
    }
    if (matchingTypes.length === 0) {
        return {
            error: `No compatible connection types found between "${sourceNode.name}" (outputs: ${sourceOutputTypes.join(', ') || 'none'}) and "${targetNode.name}" (inputs: ${targetInputTypes.join(', ') || 'none'})`,
        };
    }
    if (matchingTypes.length === 1) {
        return { connectionType: matchingTypes[0] };
    }
    return {
        possibleTypes: matchingTypes,
        error: `Multiple connection types possible: ${matchingTypes.join(', ')}. Please specify which one to use.`,
    };
}
