// @n8n/api-types/dist/dto/user/users-list-filter.dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsersListFilterDto = exports.USERS_LIST_SORT_OPTIONS = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const zod_1 = require("zod");
const zod_class_1 = require("zod-class");
const pagination_dto_1 = require("../pagination/pagination.dto");
exports.USERS_LIST_SORT_OPTIONS = [
    'firstName:asc',
    'firstName:desc',
    'lastName:asc',
    'lastName:desc',
    'email:asc',
    'email:desc',
    'role:asc',
    'role:desc',
    'mfaEnabled:asc',
    'mfaEnabled:desc',
    'lastActiveAt:asc',
    'lastActiveAt:desc',
];
const usersListSortByValidator = zod_1.z
    .array(zod_1.z.enum(exports.USERS_LIST_SORT_OPTIONS, {
    message: `sortBy must be one of: ${exports.USERS_LIST_SORT_OPTIONS.join(', ')}`,
}))
    .optional();
const userSelectSchema = zod_1.z.array(zod_1.z.enum(['id', 'firstName', 'lastName', 'email', 'disabled', 'mfaEnabled', 'role']));
const userFilterSchema = zod_1.z.object({
    isOwner: zod_1.z.boolean().optional(),
    firstName: zod_1.z.string().optional(),
    lastName: zod_1.z.string().optional(),
    email: zod_1.z.string().optional(),
    mfaEnabled: zod_1.z.boolean().optional(),
    fullText: zod_1.z.string().optional(),
});
const filterValidatorSchema = zod_1.z
    .string()
    .optional()
    .transform((val, ctx) => {
    if (!val)
        return undefined;
    try {
        const parsed = (0, n8n_workflow_1.jsonParse)(val);
        try {
            return userFilterSchema.parse(parsed);
        }
        catch (e) {
            ctx.addIssue({
                code: zod_1.z.ZodIssueCode.custom,
                message: 'Invalid filter fields',
                path: ['filter'],
            });
            return zod_1.z.NEVER;
        }
    }
    catch (e) {
        ctx.addIssue({
            code: zod_1.z.ZodIssueCode.custom,
            message: 'Invalid filter format',
            path: ['filter'],
        });
        return zod_1.z.NEVER;
    }
});
const userExpandSchema = zod_1.z.array(zod_1.z.enum(['projectRelations']));
class UsersListFilterDto extends zod_class_1.Z.class({
    ...pagination_dto_1.paginationSchema,
    take: (0, pagination_dto_1.createTakeValidator)(50, true),
    select: userSelectSchema.optional(),
    filter: filterValidatorSchema.optional(),
    expand: userExpandSchema.optional(),
    sortBy: usersListSortByValidator,
}) {
}
exports.UsersListFilterDto = UsersListFilterDto;
