// n8n-nodes-base/dist/nodes/Discord/v2/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
/* eslint-disable n8n-nodes-base/node-filename-against-convention */
const n8n_workflow_1 = require("n8n-workflow");
const channel = __importStar(require("./channel"));
const member = __importStar(require("./member"));
const message = __importStar(require("./message"));
const webhook = __importStar(require("./webhook"));
const descriptions_1 = require("../../../../utils/sendAndWait/descriptions");
exports.versionDescription = {
    displayName: 'Discord',
    name: 'discord',
    icon: 'file:discord.svg',
    group: ['output'],
    version: 2,
    subtitle: '={{ $parameter["operation"] + ": " + $parameter["resource"] }}',
    description: 'Sends data to Discord',
    defaults: {
        name: 'Discord',
    },
    inputs: [n8n_workflow_1.NodeConnectionTypes.Main],
    outputs: [n8n_workflow_1.NodeConnectionTypes.Main],
    webhooks: descriptions_1.sendAndWaitWebhooksDescription,
    credentials: [
        {
            name: 'discordBotApi',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['botToken'],
                },
            },
        },
        {
            name: 'discordOAuth2Api',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['oAuth2'],
                },
            },
        },
        {
            name: 'discordWebhookApi',
            displayOptions: {
                show: {
                    authentication: ['webhook'],
                },
            },
        },
    ],
    properties: [
        {
            displayName: 'Connection Type',
            name: 'authentication',
            type: 'options',
            options: [
                {
                    name: 'Bot Token',
                    value: 'botToken',
                    description: 'Manage messages, channels, and members on a server',
                },
                {
                    name: 'OAuth2',
                    value: 'oAuth2',
                    description: "Same features as 'Bot Token' with easier Bot installation",
                },
                {
                    name: 'Webhook',
                    value: 'webhook',
                    description: 'Send messages to a specific channel',
                },
            ],
            default: 'botToken',
        },
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'Channel',
                    value: 'channel',
                },
                {
                    name: 'Message',
                    value: 'message',
                },
                {
                    name: 'Member',
                    value: 'member',
                },
            ],
            default: 'channel',
            displayOptions: {
                hide: {
                    authentication: ['webhook'],
                },
            },
        },
        ...message.description,
        ...channel.description,
        ...member.description,
        ...webhook.description,
    ],
};
