// @n8n/n8n-nodes-langchain/dist/nodes/llms/LMOpenHuggingFaceInference/LmOpenHuggingFaceInference.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LmOpenHuggingFaceInference_node_exports = {};
__export(LmOpenHuggingFaceInference_node_exports, {
  LmOpenHuggingFaceInference: () => LmOpenHuggingFaceInference
});
module.exports = __toCommonJS(LmOpenHuggingFaceInference_node_exports);
var import_hf = require("@langchain/community/llms/hf");
var import_n8n_workflow = require("n8n-workflow");
var import_sharedFields = require("../../../utils/sharedFields");
var import_n8nLlmFailedAttemptHandler = require("../n8nLlmFailedAttemptHandler");
var import_N8nLlmTracing = require("../N8nLlmTracing");
class LmOpenHuggingFaceInference {
  constructor() {
    this.description = {
      displayName: "Hugging Face Inference Model",
      name: "lmOpenHuggingFaceInference",
      icon: "file:huggingface.svg",
      group: ["transform"],
      version: 1,
      description: "Language Model HuggingFaceInference",
      defaults: {
        name: "Hugging Face Inference Model"
      },
      codex: {
        categories: ["AI"],
        subcategories: {
          AI: ["Language Models", "Root Nodes"],
          "Language Models": ["Text Completion Models"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.lmopenhuggingfaceinference/"
            }
          ]
        }
      },
      inputs: [],
      outputs: [import_n8n_workflow.NodeConnectionTypes.AiLanguageModel],
      outputNames: ["Model"],
      credentials: [
        {
          name: "huggingFaceApi",
          required: true
        }
      ],
      properties: [
        (0, import_sharedFields.getConnectionHintNoticeField)([import_n8n_workflow.NodeConnectionTypes.AiChain, import_n8n_workflow.NodeConnectionTypes.AiAgent]),
        {
          displayName: "Model",
          name: "model",
          type: "string",
          default: "mistralai/Mistral-Nemo-Base-2407"
        },
        {
          displayName: "Options",
          name: "options",
          placeholder: "Add Option",
          description: "Additional options to add",
          type: "collection",
          default: {},
          options: [
            {
              displayName: "Custom Inference Endpoint",
              name: "endpointUrl",
              default: "",
              description: "Custom endpoint URL",
              type: "string"
            },
            {
              displayName: "Frequency Penalty",
              name: "frequencyPenalty",
              default: 0,
              typeOptions: { maxValue: 2, minValue: -2, numberPrecision: 1 },
              description: "Positive values penalize new tokens based on their existing frequency in the text so far, decreasing the model's likelihood to repeat the same line verbatim",
              type: "number"
            },
            {
              displayName: "Maximum Number of Tokens",
              name: "maxTokens",
              default: 128,
              description: "The maximum number of tokens to generate in the completion. Most models have a context length of 2048 tokens (except for the newest models, which support 32,768).",
              type: "number",
              typeOptions: {
                maxValue: 32768
              }
            },
            {
              displayName: "Presence Penalty",
              name: "presencePenalty",
              default: 0,
              typeOptions: { maxValue: 2, minValue: -2, numberPrecision: 1 },
              description: "Positive values penalize new tokens based on whether they appear in the text so far, increasing the model's likelihood to talk about new topics",
              type: "number"
            },
            {
              displayName: "Sampling Temperature",
              name: "temperature",
              default: 1,
              typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
              description: "Controls randomness: Lowering results in less random completions. As the temperature approaches zero, the model will become deterministic and repetitive.",
              type: "number"
            },
            {
              displayName: "Top K",
              name: "topK",
              default: 1,
              typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
              description: "Controls the top tokens to consider within the sample operation to create new text",
              type: "number"
            },
            {
              displayName: "Top P",
              name: "topP",
              default: 1,
              typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
              description: "Controls diversity via nucleus sampling: 0.5 means half of all likelihood-weighted options are considered. We generally recommend altering this or temperature but not both.",
              type: "number"
            }
          ]
        }
      ]
    };
  }
  async supplyData(itemIndex) {
    const credentials = await this.getCredentials("huggingFaceApi");
    const modelName = this.getNodeParameter("model", itemIndex);
    const options = this.getNodeParameter("options", itemIndex, {});
    const model = new import_hf.HuggingFaceInference({
      model: modelName,
      apiKey: credentials.apiKey,
      ...options,
      callbacks: [new import_N8nLlmTracing.N8nLlmTracing(this)],
      onFailedAttempt: (0, import_n8nLlmFailedAttemptHandler.makeN8nLlmFailedAttemptHandler)(this)
    });
    return {
      response: model
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LmOpenHuggingFaceInference
});
