// @n8n/db/dist/repositories/role.repository.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoleRepository = void 0;
const config_1 = require("@n8n/config");
const di_1 = require("@n8n/di");
const typeorm_1 = require("@n8n/typeorm");
const n8n_workflow_1 = require("n8n-workflow");
const entities_1 = require("../entities");
let RoleRepository = class RoleRepository extends typeorm_1.Repository {
    constructor(dataSource, databaseConfig) {
        super(entities_1.Role, dataSource.manager);
        this.databaseConfig = databaseConfig;
    }
    async findAll() {
        return await this.find({ relations: ['scopes'] });
    }
    async countUsersWithRole(role) {
        if (role.roleType === 'global') {
            return await this.manager.getRepository(entities_1.User).count({
                where: {
                    role: {
                        slug: role.slug,
                    },
                },
            });
        }
        else if (role.roleType === 'project') {
            return await this.manager.getRepository(entities_1.ProjectRelation).count({
                where: { role: { slug: role.slug } },
            });
        }
        return 0;
    }
    async findAllRoleCounts() {
        const userCount = await this.manager
            .createQueryBuilder(entities_1.User, 'user')
            .select('user.roleSlug', 'roleSlug')
            .addSelect('COUNT(user.id)', 'count')
            .groupBy('user.roleSlug')
            .getRawMany();
        const projectCount = await this.manager
            .createQueryBuilder(entities_1.ProjectRelation, 'projectRelation')
            .select('projectRelation.role', 'roleSlug')
            .addSelect('COUNT(projectRelation.user)', 'count')
            .groupBy('projectRelation.role')
            .getRawMany();
        return userCount.concat(projectCount).reduce((acc, { roleSlug, count }) => {
            if (!acc[roleSlug]) {
                acc[roleSlug] = 0;
            }
            acc[roleSlug] += parseInt(count, 10);
            return acc;
        }, {});
    }
    async findBySlug(slug) {
        return await this.findOne({
            where: { slug },
            relations: ['scopes'],
        });
    }
    async findBySlugs(slugs, roleType) {
        return await this.find({
            where: { slug: (0, typeorm_1.In)(slugs), roleType },
            relations: ['scopes'],
        });
    }
    async removeBySlug(slug) {
        const result = await this.delete({ slug });
        if (result.affected !== 1) {
            throw new Error(`Failed to delete role "${slug}"`);
        }
    }
    async updateEntityWithManager(entityManager, slug, newData) {
        const role = await entityManager.findOne(entities_1.Role, {
            where: { slug },
            relations: ['scopes'],
        });
        if (!role) {
            throw new n8n_workflow_1.UserError('Role not found');
        }
        if (role.systemRole) {
            throw new n8n_workflow_1.UserError('Cannot update system roles');
        }
        if (newData.displayName !== undefined) {
            role.displayName = newData.displayName;
        }
        if (newData.description !== undefined) {
            role.description = newData.description;
        }
        if (newData.scopes !== undefined) {
            role.scopes = newData.scopes;
        }
        return await entityManager.save(role);
    }
    async updateRole(slug, newData) {
        if (this.databaseConfig.isLegacySqlite) {
            return await this.updateEntityWithManager(this.manager, slug, newData);
        }
        return await this.manager.transaction(async (transactionManager) => {
            return await this.updateEntityWithManager(transactionManager, slug, newData);
        });
    }
};
exports.RoleRepository = RoleRepository;
exports.RoleRepository = RoleRepository = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [typeorm_1.DataSource,
        config_1.DatabaseConfig])
], RoleRepository);
