// @n8n/db/dist/repositories/test-run.repository.ee.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestRunRepository = void 0;
const di_1 = require("@n8n/di");
const typeorm_1 = require("@n8n/typeorm");
const n8n_workflow_1 = require("n8n-workflow");
const entities_1 = require("../entities");
const get_final_test_result_1 = require("../utils/get-final-test-result");
let TestRunRepository = class TestRunRepository extends typeorm_1.Repository {
    constructor(dataSource) {
        super(entities_1.TestRun, dataSource.manager);
    }
    async createTestRun(workflowId) {
        const testRun = this.create({
            status: 'new',
            workflow: {
                id: workflowId,
            },
        });
        return await this.save(testRun);
    }
    async markAsRunning(id) {
        return await this.update(id, {
            status: 'running',
            runAt: new Date(),
        });
    }
    async markAsCompleted(id, metrics) {
        return await this.update(id, { status: 'completed', completedAt: new Date(), metrics });
    }
    async markAsCancelled(id, trx) {
        trx = trx ?? this.manager;
        return await trx.update(entities_1.TestRun, id, { status: 'cancelled', completedAt: new Date() });
    }
    async markAsError(id, errorCode, errorDetails) {
        return await this.update(id, {
            status: 'error',
            errorCode,
            errorDetails,
            completedAt: new Date(),
        });
    }
    async markAllIncompleteAsFailed() {
        return await this.update({ status: (0, typeorm_1.In)(['new', 'running']) }, { status: 'error', errorCode: 'INTERRUPTED', completedAt: new Date() });
    }
    async getMany(workflowId, options) {
        const findManyOptions = {
            where: { workflow: { id: workflowId } },
            order: { createdAt: 'DESC' },
            relations: ['testCaseExecutions'],
        };
        if (options?.take) {
            findManyOptions.skip = options.skip;
            findManyOptions.take = options.take;
        }
        const testRuns = await this.find(findManyOptions);
        return testRuns.map(({ testCaseExecutions, ...testRun }) => {
            const finalResult = testRun.status === 'completed' ? (0, get_final_test_result_1.getTestRunFinalResult)(testCaseExecutions) : null;
            return { ...testRun, finalResult };
        });
    }
    async getTestRunSummaryById(testRunId) {
        const testRun = await this.findOne({
            where: { id: testRunId },
            relations: ['testCaseExecutions'],
        });
        if (!testRun) {
            throw new n8n_workflow_1.UnexpectedError('Test run not found');
        }
        testRun.finalResult =
            testRun.status === 'completed' ? (0, get_final_test_result_1.getTestRunFinalResult)(testRun.testCaseExecutions) : null;
        return testRun;
    }
};
exports.TestRunRepository = TestRunRepository;
exports.TestRunRepository = TestRunRepository = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [typeorm_1.DataSource])
], TestRunRepository);
