// n8n-nodes-base/dist/nodes/Microsoft/SharePoint/descriptions/item/getAll.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const utils_1 = require("../../helpers/utils");
const common_descriptions_1 = require("../common.descriptions");
const properties = [
    {
        ...common_descriptions_1.siteRLC,
        description: 'Select the site to retrieve lists from',
    },
    {
        ...common_descriptions_1.listRLC,
        description: 'Select the list you want to search for items in',
        displayOptions: {
            hide: {
                ...common_descriptions_1.untilSiteSelected,
            },
        },
    },
    {
        displayName: 'Filter by Formula',
        name: 'filter',
        default: '',
        description: 'The formula will be evaluated for each record. <a href="https://learn.microsoft.com/en-us/graph/filter-query-parameter">More info</a>.',
        hint: 'If empty, all the items will be returned',
        placeholder: "e.g. fields/Title eq 'item1'",
        routing: {
            send: {
                property: '$filter',
                type: 'query',
                value: '={{ $value ? $value : undefined }}',
            },
        },
        type: 'string',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
        routing: {
            send: {
                paginate: '={{ $value }}',
            },
            operations: {
                pagination: {
                    type: 'generic',
                    properties: {
                        continue: '={{ !!$response.body?.["@odata.nextLink"] }}',
                        request: {
                            url: '={{ $response.body?.["@odata.nextLink"] ?? $request.url }}',
                            qs: {
                                $select: '={{ !!$response.body?.["@odata.nextLink"] ? undefined : $request.qs?.$select }}',
                            },
                        },
                    },
                },
            },
        },
        type: 'boolean',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        default: 50,
        description: 'Max number of results to return',
        displayOptions: {
            show: {
                returnAll: [false],
            },
        },
        routing: {
            send: {
                property: '$top',
                type: 'query',
                value: '={{ $value }}',
            },
        },
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        validateType: 'number',
    },
    {
        displayName: 'Options',
        name: 'options',
        default: {},
        options: [
            {
                displayName: 'Fields',
                name: 'fields',
                default: [],
                description: 'The fields you want to include in the output',
                displayOptions: {
                    hide: {
                        '/simplify': [true],
                    },
                },
                options: [
                    {
                        name: 'Content Type',
                        value: 'contentType',
                    },
                    {
                        name: 'Created At',
                        value: 'createdDateTime',
                    },
                    {
                        name: 'Created By',
                        value: 'createdBy',
                    },
                    {
                        name: 'Fields',
                        value: 'fields',
                    },
                    {
                        name: 'ID',
                        value: 'id',
                    },
                    {
                        name: 'Last Modified At',
                        value: 'lastModifiedDateTime',
                    },
                    {
                        name: 'Last Modified By',
                        value: 'lastModifiedBy',
                    },
                    {
                        name: 'Parent Reference',
                        value: 'parentReference',
                    },
                    {
                        name: 'Web URL',
                        value: 'webUrl',
                    },
                ],
                routing: {
                    send: {
                        preSend: [utils_1.itemGetAllFieldsPreSend],
                    },
                },
                type: 'multiOptions',
            },
        ],
        placeholder: 'Add option',
        type: 'collection',
    },
    {
        displayName: 'Simplify',
        name: 'simplify',
        default: true,
        routing: {
            send: {
                preSend: [
                    async function (requestOptions) {
                        const simplify = this.getNodeParameter('simplify', false);
                        if (simplify) {
                            requestOptions.qs ??= {};
                            requestOptions.qs.$select = 'id,createdDateTime,lastModifiedDateTime,webUrl';
                            requestOptions.qs.$expand = 'fields(select=Title)';
                        }
                        return requestOptions;
                    },
                ],
            },
        },
        type: 'boolean',
    },
];
const displayOptions = {
    show: {
        resource: ['item'],
        operation: ['getAll'],
    },
};
exports.description = (0, n8n_workflow_1.updateDisplayOptions)(displayOptions, properties);
