// @n8n/ai-workflow-builder/dist/llm-config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.anthropicClaudeSonnet4 = exports.gpt41 = exports.gpt41mini = exports.o4mini = void 0;
const constants_1 = require("./constants");
const o4mini = async (config) => {
    const { ChatOpenAI } = await Promise.resolve().then(() => __importStar(require('@langchain/openai')));
    return new ChatOpenAI({
        model: 'o4-mini-2025-04-16',
        apiKey: config.apiKey,
        configuration: {
            baseURL: config.baseUrl,
            defaultHeaders: config.headers,
        },
    });
};
exports.o4mini = o4mini;
const gpt41mini = async (config) => {
    const { ChatOpenAI } = await Promise.resolve().then(() => __importStar(require('@langchain/openai')));
    return new ChatOpenAI({
        model: 'gpt-4.1-mini-2025-04-14',
        apiKey: config.apiKey,
        temperature: 0,
        maxTokens: -1,
        configuration: {
            baseURL: config.baseUrl,
            defaultHeaders: config.headers,
        },
    });
};
exports.gpt41mini = gpt41mini;
const gpt41 = async (config) => {
    const { ChatOpenAI } = await Promise.resolve().then(() => __importStar(require('@langchain/openai')));
    return new ChatOpenAI({
        model: 'gpt-4.1-2025-04-14',
        apiKey: config.apiKey,
        temperature: 0.3,
        maxTokens: -1,
        configuration: {
            baseURL: config.baseUrl,
            defaultHeaders: config.headers,
        },
    });
};
exports.gpt41 = gpt41;
const anthropicClaudeSonnet4 = async (config) => {
    const { ChatAnthropic } = await Promise.resolve().then(() => __importStar(require('@langchain/anthropic')));
    return new ChatAnthropic({
        model: 'claude-sonnet-4-20250514',
        apiKey: config.apiKey,
        temperature: 0,
        maxTokens: constants_1.MAX_OUTPUT_TOKENS,
        anthropicApiUrl: config.baseUrl,
        clientOptions: {
            defaultHeaders: config.headers,
        },
    });
};
exports.anthropicClaudeSonnet4 = anthropicClaudeSonnet4;
