// @n8n/ai-workflow-builder/dist/workflow-state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowState = void 0;
exports.createTrimMessagesReducer = createTrimMessagesReducer;
const messages_1 = require("@langchain/core/messages");
const langgraph_1 = require("@langchain/langgraph");
function operationsReducer(current, update) {
    if (update === null) {
        return [];
    }
    if (!update || update.length === 0) {
        return current ?? [];
    }
    if (update.some((op) => op.type === 'clear')) {
        return update.filter((op) => op.type === 'clear').slice(-1);
    }
    if (!current && !update) {
        return [];
    }
    return [...(current ?? []), ...update];
}
function createTrimMessagesReducer(maxUserMessages) {
    return (current) => {
        const humanMessageIndices = [];
        current.forEach((msg, index) => {
            if (msg instanceof messages_1.HumanMessage) {
                humanMessageIndices.push(index);
            }
        });
        if (humanMessageIndices.length <= maxUserMessages) {
            return current;
        }
        const startHumanMessageIndex = humanMessageIndices[humanMessageIndices.length - maxUserMessages];
        return current.slice(startHumanMessageIndex);
    };
}
exports.WorkflowState = langgraph_1.Annotation.Root({
    messages: (0, langgraph_1.Annotation)({
        reducer: langgraph_1.messagesStateReducer,
        default: () => [],
    }),
    workflowJSON: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => ({ nodes: [], connections: {}, name: '' }),
    }),
    workflowOperations: (0, langgraph_1.Annotation)({
        reducer: operationsReducer,
        default: () => [],
    }),
    workflowContext: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
    }),
    previousSummary: (0, langgraph_1.Annotation)({
        reducer: (x, y) => y ?? x,
        default: () => 'EMPTY',
    }),
});
