// n8n-nodes-base/dist/nodes/Google/BusinessProfile/GoogleBusinessProfileTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleBusinessProfileTrigger = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
class GoogleBusinessProfileTrigger {
    description = {
        displayName: 'Google Business Profile Trigger',
        name: 'googleBusinessProfileTrigger',
        icon: 'file:googleBusinessProfile.svg',
        group: ['trigger'],
        version: 1,
        description: 'Fetches reviews from Google Business Profile and starts the workflow on specified polling intervals.',
        subtitle: '={{"Google Business Profile Trigger"}}',
        defaults: {
            name: 'Google Business Profile Trigger',
        },
        credentials: [
            {
                name: 'googleBusinessProfileOAuth2Api',
                required: true,
            },
        ],
        polling: true,
        inputs: [],
        outputs: [n8n_workflow_1.NodeConnectionTypes.Main],
        properties: [
            {
                displayName: 'Event',
                name: 'event',
                required: true,
                type: 'options',
                noDataExpression: true,
                default: 'reviewAdded',
                options: [
                    {
                        name: 'Review Added',
                        value: 'reviewAdded',
                    },
                ],
            },
            {
                displayName: 'Account',
                name: 'account',
                required: true,
                type: 'resourceLocator',
                default: { mode: 'list', value: '' },
                description: 'The Google Business Profile account',
                displayOptions: { show: { event: ['reviewAdded'] } },
                modes: [
                    {
                        displayName: 'From list',
                        name: 'list',
                        type: 'list',
                        typeOptions: {
                            searchListMethod: 'searchAccounts',
                            searchable: true,
                        },
                    },
                    {
                        displayName: 'By name',
                        name: 'name',
                        type: 'string',
                        hint: 'Enter the account name',
                        validation: [
                            {
                                type: 'regex',
                                properties: {
                                    regex: 'accounts/[0-9]+',
                                    errorMessage: 'The name must start with "accounts/"',
                                },
                            },
                        ],
                        placeholder: 'e.g. accounts/0123456789',
                    },
                ],
            },
            {
                displayName: 'Location',
                name: 'location',
                required: true,
                type: 'resourceLocator',
                default: { mode: 'list', value: '' },
                description: 'The specific location or business associated with the account',
                displayOptions: { show: { event: ['reviewAdded'] } },
                modes: [
                    {
                        displayName: 'From list',
                        name: 'list',
                        type: 'list',
                        typeOptions: {
                            searchListMethod: 'searchLocations',
                            searchable: true,
                        },
                    },
                    {
                        displayName: 'By name',
                        name: 'name',
                        type: 'string',
                        hint: 'Enter the location name',
                        validation: [
                            {
                                type: 'regex',
                                properties: {
                                    regex: 'locations/[0-9]+',
                                    errorMessage: 'The name must start with "locations/"',
                                },
                            },
                        ],
                        placeholder: 'e.g. locations/0123456789',
                    },
                ],
            },
        ],
    };
    methods = {
        listSearch: {
            searchAccounts: GenericFunctions_1.searchAccounts,
            searchLocations: GenericFunctions_1.searchLocations,
        },
    };
    async poll() {
        const nodeStaticData = this.getWorkflowStaticData('node');
        let responseData;
        const qs = {};
        const account = this.getNodeParameter('account').value;
        const location = this.getNodeParameter('location').value;
        const manualMode = this.getMode() === 'manual';
        if (manualMode) {
            qs.pageSize = 1; // In manual mode we only want to fetch the latest review
        }
        else {
            qs.pageSize = 50; // Maximal page size for the get reviews endpoint
        }
        try {
            responseData = (await GenericFunctions_1.googleApiRequest.call(this, 'GET', `/${account}/${location}/reviews`, {}, qs));
            if (manualMode) {
                responseData = responseData.reviews;
            }
            else {
                // During the first execution there is no delta
                if (!nodeStaticData.totalReviewCountLastTimeChecked) {
                    nodeStaticData.totalReviewCountLastTimeChecked = responseData.totalReviewCount;
                    return null;
                }
                // When count did't change the node shouldn't trigger
                if (!responseData?.reviews?.length ||
                    nodeStaticData?.totalReviewCountLastTimeChecked === responseData?.totalReviewCount) {
                    return null;
                }
                const numNewReviews = 
                // @ts-ignore
                responseData.totalReviewCount - nodeStaticData.totalReviewCountLastTimeChecked;
                nodeStaticData.totalReviewCountLastTimeChecked = responseData.totalReviewCount;
                // By default the reviews will be sorted by updateTime in descending order
                // Return only the delta reviews since last pooling
                responseData = responseData.reviews.slice(0, numNewReviews);
            }
            if (Array.isArray(responseData) && responseData.length) {
                return [this.helpers.returnJsonArray(responseData)];
            }
            return null;
        }
        catch (error) {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
        }
    }
}
exports.GoogleBusinessProfileTrigger = GoogleBusinessProfileTrigger;
