// @n8n/n8n-nodes-langchain/dist/nodes/vendors/GoogleGemini/actions/file/upload.operation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var upload_operation_exports = {};
__export(upload_operation_exports, {
  description: () => description,
  execute: () => execute,
  properties: () => properties
});
module.exports = __toCommonJS(upload_operation_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_utils = require("../../helpers/utils");
const properties = [
  {
    displayName: "Input Type",
    name: "inputType",
    type: "options",
    default: "url",
    options: [
      {
        name: "File URL",
        value: "url"
      },
      {
        name: "Binary File",
        value: "binary"
      }
    ]
  },
  {
    displayName: "URL",
    name: "fileUrl",
    type: "string",
    placeholder: "e.g. https://example.com/file.pdf",
    description: "URL of the file to upload",
    default: "",
    displayOptions: {
      show: {
        inputType: ["url"]
      }
    }
  },
  {
    displayName: "Input Data Field Name",
    name: "binaryPropertyName",
    type: "string",
    default: "data",
    placeholder: "e.g. data",
    hint: "The name of the input field containing the binary file data to be processed",
    description: "Name of the binary property which contains the file",
    displayOptions: {
      show: {
        inputType: ["binary"]
      }
    }
  }
];
const displayOptions = {
  show: {
    operation: ["upload"],
    resource: ["file"]
  }
};
const description = (0, import_n8n_workflow.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
  const inputType = this.getNodeParameter("inputType", i, "url");
  let fileUrl;
  if (inputType === "url") {
    fileUrl = this.getNodeParameter("fileUrl", i, "");
  }
  const response = await import_utils.transferFile.call(this, i, fileUrl, "application/octet-stream");
  return [
    {
      json: response,
      pairedItem: {
        item: i
      }
    }
  ];
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  description,
  execute,
  properties
});
