// @n8n/backend-test-utils/dist/db/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllProjectRelations = exports.getProjectRoleForUser = exports.getProjectRelations = exports.findProject = exports.getPersonalProject = exports.createTeamProject = exports.linkUserToProject = void 0;
exports.getProjectByNameOrFail = getProjectByNameOrFail;
const db_1 = require("@n8n/db");
const di_1 = require("@n8n/di");
const permissions_1 = require("@n8n/permissions");
const random_1 = require("../random");
const linkUserToProject = async (user, project, role) => {
    const projectRelationRepository = di_1.Container.get(db_1.ProjectRelationRepository);
    await projectRelationRepository.save(projectRelationRepository.create({
        projectId: project.id,
        userId: user.id,
        role: { slug: role },
    }));
};
exports.linkUserToProject = linkUserToProject;
const createTeamProject = async (name, adminUser) => {
    const projectRepository = di_1.Container.get(db_1.ProjectRepository);
    const project = await projectRepository.save(projectRepository.create({
        name: name ?? (0, random_1.randomName)(),
        type: 'team',
    }));
    if (adminUser) {
        await (0, exports.linkUserToProject)(adminUser, project, 'project:admin');
    }
    return project;
};
exports.createTeamProject = createTeamProject;
async function getProjectByNameOrFail(name) {
    return await di_1.Container.get(db_1.ProjectRepository).findOneOrFail({ where: { name } });
}
const getPersonalProject = async (user) => {
    return await di_1.Container.get(db_1.ProjectRepository).findOneOrFail({
        where: {
            projectRelations: {
                userId: user.id,
                role: { slug: permissions_1.PROJECT_OWNER_ROLE_SLUG },
            },
            type: 'personal',
        },
    });
};
exports.getPersonalProject = getPersonalProject;
const findProject = async (id) => {
    return await di_1.Container.get(db_1.ProjectRepository).findOneOrFail({
        where: { id },
    });
};
exports.findProject = findProject;
const getProjectRelations = async ({ projectId, userId, role, }) => {
    return await di_1.Container.get(db_1.ProjectRelationRepository).find({
        where: { projectId, userId, role },
        relations: { role: true },
    });
};
exports.getProjectRelations = getProjectRelations;
const getProjectRoleForUser = async (projectId, userId) => {
    return (await di_1.Container.get(db_1.ProjectRelationRepository).findOne({
        where: { projectId, userId },
        relations: { role: true },
    }))?.role?.slug;
};
exports.getProjectRoleForUser = getProjectRoleForUser;
const getAllProjectRelations = async ({ projectId, }) => {
    return await di_1.Container.get(db_1.ProjectRelationRepository).find({
        where: { projectId },
        relations: { role: true },
    });
};
exports.getAllProjectRelations = getAllProjectRelations;
