// n8n-nodes-base/dist/nodes/TheHiveProject/actions/task/search.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = void 0;
exports.execute = execute;
const utilities_1 = require("../../../../utils/utilities");
const descriptions_1 = require("../../descriptions");
const transport_1 = require("../../transport");
const properties = [
    {
        displayName: 'Search in All Cases',
        name: 'allCases',
        type: 'boolean',
        default: true,
        description: 'Whether to search in all cases or only in a selected case',
    },
    {
        ...descriptions_1.caseRLC,
        displayOptions: {
            show: {
                allCases: [false],
            },
        },
    },
    ...descriptions_1.returnAllAndLimit,
    descriptions_1.genericFiltersCollection,
    descriptions_1.sortCollection,
    descriptions_1.searchOptions,
];
const displayOptions = {
    show: {
        resource: ['task'],
        operation: ['search'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(i) {
    let responseData = [];
    const allCases = this.getNodeParameter('allCases', i);
    const filtersValues = this.getNodeParameter('filters.values', i, []);
    const sortFields = this.getNodeParameter('sort.fields', i, []);
    const returnAll = this.getNodeParameter('returnAll', i);
    const { returnCount, extraData } = this.getNodeParameter('options', i);
    let limit;
    let scope;
    if (allCases) {
        scope = { query: 'listTask' };
    }
    else {
        const caseId = this.getNodeParameter('caseId', i, '', { extractValue: true });
        scope = { query: 'getCase', id: caseId, restrictTo: 'tasks' };
    }
    if (!returnAll) {
        limit = this.getNodeParameter('limit', i);
    }
    responseData = await transport_1.theHiveApiQuery.call(this, scope, filtersValues, sortFields, limit, returnCount, extraData);
    const executionData = this.helpers.constructExecutionMetaData((0, utilities_1.wrapData)(responseData), {
        itemData: { item: i },
    });
    return executionData;
}
