// n8n-nodes-base/dist/nodes/Microsoft/Outlook/v2/actions/draft/send.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.properties = void 0;
exports.execute = execute;
const utilities_1 = require("../../../../../../utils/utilities");
const descriptions_1 = require("../../descriptions");
const utils_1 = require("../../helpers/utils");
const transport_1 = require("../../transport");
exports.properties = [
    descriptions_1.draftRLC,
    {
        displayName: 'To',
        name: 'to',
        description: 'Comma-separated list of email addresses of recipients',
        type: 'string',
        default: '',
    },
];
const displayOptions = {
    show: {
        resource: ['draft'],
        operation: ['send'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, exports.properties);
async function execute(index) {
    const draftId = this.getNodeParameter('draftId', index, undefined, { extractValue: true });
    const to = this.getNodeParameter('to', index);
    if (to) {
        const recipients = to
            .split(',')
            .map((s) => s.trim())
            .filter((email) => email);
        if (recipients.length !== 0) {
            await transport_1.microsoftApiRequest.call(this, 'PATCH', `/messages/${draftId}`, {
                toRecipients: recipients.map((recipient) => (0, utils_1.makeRecipient)(recipient)),
            });
        }
    }
    await transport_1.microsoftApiRequest.call(this, 'POST', `/messages/${draftId}/send`);
    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: index } });
    return executionData;
}
