// @n8n/n8n-nodes-langchain/dist/nodes/vector_store/VectorStoreZepInsert/VectorStoreZepInsert.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var VectorStoreZepInsert_node_exports = {};
__export(VectorStoreZepInsert_node_exports, {
  VectorStoreZepInsert: () => VectorStoreZepInsert
});
module.exports = __toCommonJS(VectorStoreZepInsert_node_exports);
var import_zep = require("@langchain/community/vectorstores/zep");
var import_n8n_workflow = require("n8n-workflow");
var import_processDocuments = require("../shared/processDocuments");
class VectorStoreZepInsert {
  constructor() {
    this.description = {
      displayName: "Zep Vector Store: Insert",
      name: "vectorStoreZepInsert",
      hidden: true,
      // eslint-disable-next-line n8n-nodes-base/node-class-description-icon-not-svg
      icon: "file:zep.png",
      group: ["transform"],
      version: 1,
      description: "Insert data into Zep Vector Store index",
      defaults: {
        name: "Zep: Insert"
      },
      codex: {
        categories: ["AI"],
        subcategories: {
          AI: ["Vector Stores"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/cluster-nodes/root-nodes/n8n-nodes-langchain.vectorstorezep/"
            }
          ]
        }
      },
      credentials: [
        {
          name: "zepApi",
          required: true
        }
      ],
      inputs: [
        import_n8n_workflow.NodeConnectionTypes.Main,
        {
          displayName: "Document",
          maxConnections: 1,
          type: import_n8n_workflow.NodeConnectionTypes.AiDocument,
          required: true
        },
        {
          displayName: "Embedding",
          maxConnections: 1,
          type: import_n8n_workflow.NodeConnectionTypes.AiEmbedding,
          required: true
        }
      ],
      outputs: [import_n8n_workflow.NodeConnectionTypes.Main],
      properties: [
        {
          displayName: "This Zep integration is deprecated and will be removed in a future version.",
          name: "deprecationNotice",
          type: "notice",
          default: ""
        },
        {
          displayName: "Collection Name",
          name: "collectionName",
          type: "string",
          default: "",
          required: true
        },
        {
          displayName: "Specify the document to load in the document loader sub-node",
          name: "notice",
          type: "notice",
          default: ""
        },
        {
          displayName: "Options",
          name: "options",
          type: "collection",
          placeholder: "Add Option",
          default: {},
          options: [
            {
              displayName: "Embedding Dimensions",
              name: "embeddingDimensions",
              type: "number",
              default: 1536,
              description: "Whether to allow using characters from the Unicode surrogate blocks"
            },
            {
              displayName: "Is Auto Embedded",
              name: "isAutoEmbedded",
              type: "boolean",
              default: true,
              description: "Whether to automatically embed documents when they are added"
            }
          ]
        }
      ]
    };
  }
  async execute() {
    this.logger.debug("Executing data for Zep Insert Vector Store");
    const items = this.getInputData(0);
    const collectionName = this.getNodeParameter("collectionName", 0);
    const options = this.getNodeParameter("options", 0, {}) || {};
    const credentials = await this.getCredentials("zepApi");
    const documentInput = await this.getInputConnectionData(import_n8n_workflow.NodeConnectionTypes.AiDocument, 0);
    const embeddings = await this.getInputConnectionData(
      import_n8n_workflow.NodeConnectionTypes.AiEmbedding,
      0
    );
    const { processedDocuments, serializedDocuments } = await (0, import_processDocuments.processDocuments)(
      documentInput,
      items
    );
    const zepConfig = {
      apiUrl: credentials.apiUrl,
      apiKey: credentials.apiKey,
      collectionName,
      embeddingDimensions: options.embeddingDimensions ?? 1536,
      isAutoEmbedded: options.isAutoEmbedded ?? true
    };
    await import_zep.ZepVectorStore.fromDocuments(processedDocuments, embeddings, zepConfig);
    return [serializedDocuments];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  VectorStoreZepInsert
});
