// @n8n/api-types/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataStoreColumnNameSchema = exports.dateTimeSchema = exports.DATA_STORE_COLUMN_ERROR_MESSAGE = exports.DATA_STORE_COLUMN_MAX_LENGTH = exports.DATA_STORE_COLUMN_REGEX = exports.usersListSchema = exports.ROLE = exports.SOURCE_CONTROL_FILE_TYPE = exports.SOURCE_CONTROL_FILE_STATUS = exports.SOURCE_CONTROL_FILE_LOCATION = exports.passwordSchema = exports.ViewableMimeTypes = exports.heartbeatMessageSchema = exports.createHeartbeatMessage = void 0;
__exportStar(require("./dto"), exports);
var heartbeat_1 = require("./push/heartbeat");
Object.defineProperty(exports, "createHeartbeatMessage", { enumerable: true, get: function () { return heartbeat_1.createHeartbeatMessage; } });
Object.defineProperty(exports, "heartbeatMessageSchema", { enumerable: true, get: function () { return heartbeat_1.heartbeatMessageSchema; } });
var binary_data_schema_1 = require("./schemas/binary-data.schema");
Object.defineProperty(exports, "ViewableMimeTypes", { enumerable: true, get: function () { return binary_data_schema_1.ViewableMimeTypes; } });
var password_schema_1 = require("./schemas/password.schema");
Object.defineProperty(exports, "passwordSchema", { enumerable: true, get: function () { return password_schema_1.passwordSchema; } });
var source_controlled_file_schema_1 = require("./schemas/source-controlled-file.schema");
Object.defineProperty(exports, "SOURCE_CONTROL_FILE_LOCATION", { enumerable: true, get: function () { return source_controlled_file_schema_1.SOURCE_CONTROL_FILE_LOCATION; } });
Object.defineProperty(exports, "SOURCE_CONTROL_FILE_STATUS", { enumerable: true, get: function () { return source_controlled_file_schema_1.SOURCE_CONTROL_FILE_STATUS; } });
Object.defineProperty(exports, "SOURCE_CONTROL_FILE_TYPE", { enumerable: true, get: function () { return source_controlled_file_schema_1.SOURCE_CONTROL_FILE_TYPE; } });
var user_schema_1 = require("./schemas/user.schema");
Object.defineProperty(exports, "ROLE", { enumerable: true, get: function () { return user_schema_1.ROLE; } });
Object.defineProperty(exports, "usersListSchema", { enumerable: true, get: function () { return user_schema_1.usersListSchema; } });
var data_store_schema_1 = require("./schemas/data-store.schema");
Object.defineProperty(exports, "DATA_STORE_COLUMN_REGEX", { enumerable: true, get: function () { return data_store_schema_1.DATA_STORE_COLUMN_REGEX; } });
Object.defineProperty(exports, "DATA_STORE_COLUMN_MAX_LENGTH", { enumerable: true, get: function () { return data_store_schema_1.DATA_STORE_COLUMN_MAX_LENGTH; } });
Object.defineProperty(exports, "DATA_STORE_COLUMN_ERROR_MESSAGE", { enumerable: true, get: function () { return data_store_schema_1.DATA_STORE_COLUMN_ERROR_MESSAGE; } });
Object.defineProperty(exports, "dateTimeSchema", { enumerable: true, get: function () { return data_store_schema_1.dateTimeSchema; } });
Object.defineProperty(exports, "dataStoreColumnNameSchema", { enumerable: true, get: function () { return data_store_schema_1.dataStoreColumnNameSchema; } });
