// n8n-nodes-base/dist/nodes/Webhook/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkResponseModeConfiguration = exports.isIpWhitelisted = exports.setupOutputConnection = exports.configuredOutputs = exports.getResponseData = exports.getResponseCode = void 0;
exports.validateWebhookAuthentication = validateWebhookAuthentication;
const basic_auth_1 = __importDefault(require("basic-auth"));
const jsonwebtoken_1 = __importDefault(require("jsonwebtoken"));
const n8n_workflow_1 = require("n8n-workflow");
const error_1 = require("./error");
const utilities_1 = require("../../utils/utilities");
const getResponseCode = (parameters) => {
    if (parameters.responseCode) {
        return parameters.responseCode;
    }
    const responseCodeOptions = parameters.options;
    if (responseCodeOptions?.responseCode?.values) {
        const { responseCode, customCode } = responseCodeOptions.responseCode.values;
        if (customCode) {
            return customCode;
        }
        return responseCode;
    }
    return 200;
};
exports.getResponseCode = getResponseCode;
const getResponseData = (parameters) => {
    const { responseData, responseMode, options } = parameters;
    if (responseData)
        return responseData;
    if (responseMode === 'onReceived') {
        const data = options?.responseData;
        if (data)
            return data;
    }
    if (options?.noResponseBody)
        return 'noData';
    return undefined;
};
exports.getResponseData = getResponseData;
const configuredOutputs = (parameters) => {
    const httpMethod = parameters.httpMethod;
    if (!Array.isArray(httpMethod))
        return [
            {
                type: 'main',
                displayName: httpMethod,
            },
        ];
    const outputs = httpMethod.map((method) => {
        return {
            type: 'main',
            displayName: method,
        };
    });
    return outputs;
};
exports.configuredOutputs = configuredOutputs;
const setupOutputConnection = (ctx, method, additionalData) => {
    const httpMethod = ctx.getNodeParameter('httpMethod', []);
    let webhookUrl = ctx.getNodeWebhookUrl('default');
    const executionMode = ctx.getMode() === 'manual' ? 'test' : 'production';
    if (executionMode === 'test') {
        webhookUrl = webhookUrl.replace('/webhook/', '/webhook-test/');
    }
    // multi methods could be set in settings of node, so we need to check if it's an array
    if (!Array.isArray(httpMethod)) {
        return (outputData) => {
            outputData.json.webhookUrl = webhookUrl;
            outputData.json.executionMode = executionMode;
            if (additionalData?.jwtPayload) {
                outputData.json.jwtPayload = additionalData.jwtPayload;
            }
            return [[outputData]];
        };
    }
    const outputIndex = httpMethod.indexOf(method.toUpperCase());
    const outputs = httpMethod.map(() => []);
    return (outputData) => {
        outputData.json.webhookUrl = webhookUrl;
        outputData.json.executionMode = executionMode;
        if (additionalData?.jwtPayload) {
            outputData.json.jwtPayload = additionalData.jwtPayload;
        }
        outputs[outputIndex] = [outputData];
        return outputs;
    };
};
exports.setupOutputConnection = setupOutputConnection;
const isIpWhitelisted = (whitelist, ips, ip) => {
    if (whitelist === undefined || whitelist === '') {
        return true;
    }
    if (!Array.isArray(whitelist)) {
        whitelist = whitelist.split(',').map((entry) => entry.trim());
    }
    for (const address of whitelist) {
        if (ip?.includes(address)) {
            return true;
        }
        if (ips.some((entry) => entry.includes(address))) {
            return true;
        }
    }
    return false;
};
exports.isIpWhitelisted = isIpWhitelisted;
const checkResponseModeConfiguration = (context) => {
    const responseMode = context.getNodeParameter('responseMode', 'onReceived');
    const connectedNodes = context.getChildNodes(context.getNode().name);
    const isRespondToWebhookConnected = connectedNodes.some((node) => node.type === 'n8n-nodes-base.respondToWebhook');
    if (!isRespondToWebhookConnected && responseMode === 'responseNode') {
        throw new n8n_workflow_1.WorkflowConfigurationError(context.getNode(), new Error('No Respond to Webhook node found in the workflow'), {
            description: 'Insert a Respond to Webhook node to your workflow to respond to the webhook or choose another option for the “Respond” parameter',
        });
    }
    if (isRespondToWebhookConnected && !['responseNode', 'streaming'].includes(responseMode)) {
        throw new n8n_workflow_1.WorkflowConfigurationError(context.getNode(), new Error('Unused Respond to Webhook node found in the workflow'), {
            description: 'Set the “Respond” parameter to “Using Respond to Webhook Node” or remove the Respond to Webhook node',
        });
    }
};
exports.checkResponseModeConfiguration = checkResponseModeConfiguration;
async function validateWebhookAuthentication(ctx, authPropertyName) {
    const authentication = ctx.getNodeParameter(authPropertyName);
    if (authentication === 'none')
        return;
    const req = ctx.getRequestObject();
    const headers = ctx.getHeaderData();
    if (authentication === 'basicAuth') {
        // Basic authorization is needed to call webhook
        let expectedAuth;
        try {
            expectedAuth = await ctx.getCredentials('httpBasicAuth');
        }
        catch { }
        if (expectedAuth === undefined || !expectedAuth.user || !expectedAuth.password) {
            // Data is not defined on node so can not authenticate
            throw new error_1.WebhookAuthorizationError(500, 'No authentication data defined on node!');
        }
        const providedAuth = (0, basic_auth_1.default)(req);
        // Authorization data is missing
        if (!providedAuth)
            throw new error_1.WebhookAuthorizationError(401);
        if (providedAuth.name !== expectedAuth.user || providedAuth.pass !== expectedAuth.password) {
            // Provided authentication data is wrong
            throw new error_1.WebhookAuthorizationError(403);
        }
    }
    else if (authentication === 'bearerAuth') {
        let expectedAuth;
        try {
            expectedAuth = await ctx.getCredentials('httpBearerAuth');
        }
        catch { }
        const expectedToken = expectedAuth?.token;
        if (!expectedToken) {
            throw new error_1.WebhookAuthorizationError(500, 'No authentication data defined on node!');
        }
        if (headers.authorization !== `Bearer ${expectedToken}`) {
            throw new error_1.WebhookAuthorizationError(403);
        }
    }
    else if (authentication === 'headerAuth') {
        // Special header with value is needed to call webhook
        let expectedAuth;
        try {
            expectedAuth = await ctx.getCredentials('httpHeaderAuth');
        }
        catch { }
        if (expectedAuth === undefined || !expectedAuth.name || !expectedAuth.value) {
            // Data is not defined on node so can not authenticate
            throw new error_1.WebhookAuthorizationError(500, 'No authentication data defined on node!');
        }
        const headerName = expectedAuth.name.toLowerCase();
        const expectedValue = expectedAuth.value;
        if (!headers.hasOwnProperty(headerName) ||
            headers[headerName] !== expectedValue) {
            // Provided authentication data is wrong
            throw new error_1.WebhookAuthorizationError(403);
        }
    }
    else if (authentication === 'jwtAuth') {
        let expectedAuth;
        try {
            expectedAuth = await ctx.getCredentials('jwtAuth');
        }
        catch { }
        if (expectedAuth === undefined) {
            // Data is not defined on node so can not authenticate
            throw new error_1.WebhookAuthorizationError(500, 'No authentication data defined on node!');
        }
        const authHeader = req.headers.authorization;
        const token = authHeader?.split(' ')[1];
        if (!token) {
            throw new error_1.WebhookAuthorizationError(401, 'No token provided');
        }
        let secretOrPublicKey;
        if (expectedAuth.keyType === 'passphrase') {
            secretOrPublicKey = expectedAuth.secret;
        }
        else {
            secretOrPublicKey = (0, utilities_1.formatPrivateKey)(expectedAuth.publicKey, true);
        }
        try {
            return jsonwebtoken_1.default.verify(token, secretOrPublicKey, {
                algorithms: [expectedAuth.algorithm],
            });
        }
        catch (error) {
            throw new error_1.WebhookAuthorizationError(403, error.message);
        }
    }
}
