// n8n-nodes-base/dist/nodes/Google/Drive/v2/helpers/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DRIVE = exports.RLC_FOLDER_DEFAULT = exports.RLC_DRIVE_DEFAULT = exports.UPLOAD_CHUNK_SIZE = void 0;
exports.UPLOAD_CHUNK_SIZE = 256 * 1024;
exports.RLC_DRIVE_DEFAULT = 'My Drive';
exports.RLC_FOLDER_DEFAULT = 'root';
exports.DRIVE = {
    FOLDER: 'application/vnd.google-apps.folder',
    AUDIO: 'application/vnd.google-apps.audio',
    DOCUMENT: 'application/vnd.google-apps.document',
    SDK: 'application/vnd.google-apps.drive-sdk',
    DRAWING: 'application/vnd.google-apps.drawing',
    FILE: 'application/vnd.google-apps.file',
    FORM: 'application/vnd.google-apps.form',
    FUSIONTABLE: 'application/vnd.google-apps.fusiontable',
    MAP: 'application/vnd.google-apps.map',
    PHOTO: 'application/vnd.google-apps.photo',
    PRESENTATION: 'application/vnd.google-apps.presentation',
    APP_SCRIPTS: 'application/vnd.google-apps.script',
    SITES: 'application/vnd.google-apps.sites',
    SPREADSHEET: 'application/vnd.google-apps.spreadsheet',
    UNKNOWN: 'application/vnd.google-apps.unknown',
    VIDEO: 'application/vnd.google-apps.video',
};
