// @n8n/n8n-nodes-langchain/dist/nodes/mcp/McpClientTool/descriptions.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var descriptions_exports = {};
__export(descriptions_exports, {
  transportSelect: () => transportSelect
});
module.exports = __toCommonJS(descriptions_exports);
const transportSelect = ({
  defaultOption,
  displayOptions
}) => ({
  displayName: "Server Transport",
  name: "serverTransport",
  type: "options",
  options: [
    {
      name: "HTTP Streamable",
      value: "httpStreamable"
    },
    {
      name: "Server Sent Events (Deprecated)",
      value: "sse"
    }
  ],
  default: defaultOption,
  description: "The transport used by your endpoint",
  displayOptions
});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  transportSelect
});
